<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.306">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="phonchi">
<meta name="dcterms.date" content="2022-12-05">

<title>Statistical Learning and Data Mining - 10&nbsp; Unsupervised Learning</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./NumPy_tutorial.html" rel="next">
<link href="./Chapter_9_Lab.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Chapter_12_Lab.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Unsupervised Learning</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Statistical Learning and Data Mining</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter1_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter 2_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Statistical Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_3_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Regression</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_4_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Classification</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_5_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Resampling methods</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_6_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Linear Models and Regularization Methods</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_7_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Non-linear Modeling</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_8_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Decision Trees</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_9_Lab.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Support Vector Machines</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter_12_Lab.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Unsupervised Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./NumPy_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Numpy - multidimensional data arrays for python</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Colab_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Introduction to Colab</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./kaggle-explore.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Introduction to Kaggle</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#dimensional-reduction" id="toc-dimensional-reduction" class="nav-link active" data-scroll-target="#dimensional-reduction"><span class="header-section-number">10.1</span> Dimensional reduction</a>
  <ul class="collapse">
  <li><a href="#principal-components-analysis" id="toc-principal-components-analysis" class="nav-link" data-scroll-target="#principal-components-analysis"><span class="header-section-number">10.1.1</span> Principal Components Analysis</a></li>
  <li><a href="#matrix-completion" id="toc-matrix-completion" class="nav-link" data-scroll-target="#matrix-completion"><span class="header-section-number">10.1.2</span> Matrix Completion</a></li>
  <li><a href="#t-sne" id="toc-t-sne" class="nav-link" data-scroll-target="#t-sne"><span class="header-section-number">10.1.3</span> t-SNE</a></li>
  <li><a href="#umap" id="toc-umap" class="nav-link" data-scroll-target="#umap"><span class="header-section-number">10.1.4</span> UMAP</a></li>
  </ul></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span class="header-section-number">10.2</span> Clustering</a>
  <ul class="collapse">
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering"><span class="header-section-number">10.2.1</span> <span class="math inline">\(K\)</span>-Means Clustering</a></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering"><span class="header-section-number">10.2.2</span> Hierarchical Clustering</a></li>
  <li><a href="#dbscan" id="toc-dbscan" class="nav-link" data-scroll-target="#dbscan"><span class="header-section-number">10.2.3</span> DBSCAN</a></li>
  </ul></li>
  <li><a href="#nci60-data-example" id="toc-nci60-data-example" class="nav-link" data-scroll-target="#nci60-data-example"><span class="header-section-number">10.3</span> NCI60 Data Example</a>
  <ul class="collapse">
  <li><a href="#pca-on-the-nci60-data" id="toc-pca-on-the-nci60-data" class="nav-link" data-scroll-target="#pca-on-the-nci60-data"><span class="header-section-number">10.3.1</span> PCA on the NCI60 Data</a></li>
  <li><a href="#clustering-the-observations-of-the-nci60-data" id="toc-clustering-the-observations-of-the-nci60-data" class="nav-link" data-scroll-target="#clustering-the-observations-of-the-nci60-data"><span class="header-section-number">10.3.2</span> Clustering the Observations of the NCI60 Data</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Unsupervised Learning</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>phonchi </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 5, 2022</p>
    </div>
  </div>
  
    
  </div>
  

</header>


<table align="left">
<tbody><tr><td>
<a href="https://colab.research.google.com/github/phonchi/nsysu-math524/blob/master/static_files/presentations/Chapter_12_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
</td>
<td>
<a target="_blank" href="https://kaggle.com/kernels/welcome?src=https://github.com/phonchi/nsysu-math524/blob/master/static_files/presentations/Chapter_12_Lab.ipynb"><img src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>
</td>

</tr></tbody></table>
<p><br></p>
<section id="dimensional-reduction" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="dimensional-reduction"><span class="header-section-number">10.1</span> Dimensional reduction</h2>
<section id="principal-components-analysis" class="level3" data-number="10.1.1">
<h3 data-number="10.1.1" class="anchored" data-anchor-id="principal-components-analysis"><span class="header-section-number">10.1.1</span> Principal Components Analysis</h3>
<p>In this lab, we perform PCA on the <code>USArrests</code> data set. The rows of the data set contain the 50 states, in alphabetical order.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:42959,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901272414,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="a6147e33-0424-45f7-abac-354e78b7485c" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="op">!</span>pip install fancyimpute <span class="op">-</span>qq</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="op">!</span>pip install opentsne <span class="op">-</span>qq</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="op">!</span>pip install umap<span class="op">-</span>learn <span class="op">-</span>qq</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="op">!</span>pip install git<span class="op">+</span>https:<span class="op">//</span>github.com<span class="op">/</span>dmuellner<span class="op">/</span>fastcluster <span class="op">-</span>qq</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="op">!</span>pip install hdbscan <span class="op">-</span>qq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:317,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901474196,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="51">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="im">from</span> itertools <span class="im">import</span> cycle, islice</span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="im">from</span> sklearn.datasets <span class="im">import</span> load_digits</span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> scale</span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="im">from</span> sklearn.utils <span class="im">import</span> resample</span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> StandardScaler, LabelEncoder</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> train_test_split</span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="im">from</span> sklearn.svm <span class="im">import</span> SVC</span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="im">from</span> sklearn.neighbors <span class="im">import</span> KNeighborsClassifier</span>
<span id="cb3-13"><a href="#cb3-13"></a></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="im">from</span> sklearn.manifold <span class="im">import</span> TSNE</span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="im">from</span> sklearn.datasets <span class="im">import</span> make_blobs</span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="im">from</span> sklearn.cluster <span class="im">import</span> KMeans, AgglomerativeClustering, DBSCAN</span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="im">from</span> sklearn <span class="im">import</span> metrics </span>
<span id="cb3-18"><a href="#cb3-18"></a></span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="im">from</span> scipy <span class="im">import</span> linalg</span>
<span id="cb3-20"><a href="#cb3-20"></a><span class="im">from</span> scipy.cluster.hierarchy <span class="im">import</span> linkage, dendrogram, cut_tree</span>
<span id="cb3-21"><a href="#cb3-21"></a></span>
<span id="cb3-22"><a href="#cb3-22"></a><span class="im">from</span> fancyimpute <span class="im">import</span> SoftImpute <span class="co">#pip install fancyimpute</span></span>
<span id="cb3-23"><a href="#cb3-23"></a><span class="im">import</span> fastcluster <span class="im">as</span> fc  <span class="co">#pip install git+https://github.com/dmuellner/fastcluster</span></span>
<span id="cb3-24"><a href="#cb3-24"></a><span class="im">from</span> openTSNE <span class="im">import</span> TSNE <span class="im">as</span> oTSNE <span class="co">#pip install opentsne</span></span>
<span id="cb3-25"><a href="#cb3-25"></a><span class="im">import</span> hdbscan <span class="co">#pip install hdbscan</span></span>
<span id="cb3-26"><a href="#cb3-26"></a><span class="im">import</span> umap </span>
<span id="cb3-27"><a href="#cb3-27"></a></span>
<span id="cb3-28"><a href="#cb3-28"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb3-29"><a href="#cb3-29"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb3-30"><a href="#cb3-30"></a></span>
<span id="cb3-31"><a href="#cb3-31"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb3-32"><a href="#cb3-32"></a>plt.style.use(<span class="st">'seaborn-white'</span>)</span>
<span id="cb3-33"><a href="#cb3-33"></a>sns.set_context(<span class="st">"notebook"</span>, font_scale<span class="op">=</span><span class="fl">1.5</span>, rc<span class="op">=</span>{<span class="st">"lines.linewidth"</span>: <span class="fl">2.5</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901308985,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">#from https://scikit-learn.org/stable/auto_examples/cluster/plot_agglomerative_dendrogram.html</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">def</span> plot_dendrogram(model, <span class="op">**</span>kwargs):</span>
<span id="cb4-3"><a href="#cb4-3"></a>    <span class="co"># Create linkage matrix and then plot the dendrogram</span></span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a>    <span class="co"># create the counts of samples under each node</span></span>
<span id="cb4-6"><a href="#cb4-6"></a>    counts <span class="op">=</span> np.zeros(model.children_.shape[<span class="dv">0</span>])</span>
<span id="cb4-7"><a href="#cb4-7"></a>    n_samples <span class="op">=</span> <span class="bu">len</span>(model.labels_)</span>
<span id="cb4-8"><a href="#cb4-8"></a>    <span class="cf">for</span> i, merge <span class="kw">in</span> <span class="bu">enumerate</span>(model.children_):</span>
<span id="cb4-9"><a href="#cb4-9"></a>        current_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb4-10"><a href="#cb4-10"></a>        <span class="cf">for</span> child_idx <span class="kw">in</span> merge:</span>
<span id="cb4-11"><a href="#cb4-11"></a>            <span class="cf">if</span> child_idx <span class="op">&lt;</span> n_samples:</span>
<span id="cb4-12"><a href="#cb4-12"></a>                current_count <span class="op">+=</span> <span class="dv">1</span>  <span class="co"># leaf node</span></span>
<span id="cb4-13"><a href="#cb4-13"></a>            <span class="cf">else</span>:</span>
<span id="cb4-14"><a href="#cb4-14"></a>                current_count <span class="op">+=</span> counts[child_idx <span class="op">-</span> n_samples]</span>
<span id="cb4-15"><a href="#cb4-15"></a>        counts[i] <span class="op">=</span> current_count</span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a>    linkage_matrix <span class="op">=</span> np.column_stack(</span>
<span id="cb4-18"><a href="#cb4-18"></a>        [model.children_, model.distances_, counts]</span>
<span id="cb4-19"><a href="#cb4-19"></a>    ).astype(<span class="bu">float</span>)</span>
<span id="cb4-20"><a href="#cb4-20"></a></span>
<span id="cb4-21"><a href="#cb4-21"></a>    <span class="co"># Plot the corresponding dendrogram</span></span>
<span id="cb4-22"><a href="#cb4-22"></a>    dendrogram(linkage_matrix, <span class="op">**</span>kwargs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:18449,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901327430,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="e4a9311c-7ca5-4893-da8a-e03fa59525b1" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a><span class="im">from</span> google.colab <span class="im">import</span> drive</span>
<span id="cb5-2"><a href="#cb5-2"></a>drive.mount(<span class="st">'/content/drive'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Mounted at /content/drive</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1067,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328495,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="9228cef9-4982-49d3-f130-4de67043271d" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">#load the data </span></span>
<span id="cb7-2"><a href="#cb7-2"></a>USArrests <span class="op">=</span> pd.read_csv(<span class="st">'/content/drive/MyDrive/Lab/Data/USArrests.csv'</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="bu">print</span>(USArrests.shape)</span>
<span id="cb7-4"><a href="#cb7-4"></a>USArrests.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(50, 4)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="6">

  <div id="df-96937f24-dc9f-4ad9-a08b-eb8dc0f3e511">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Murder</th>
<th data-quarto-table-cell-role="th">Assault</th>
<th data-quarto-table-cell-role="th">UrbanPop</th>
<th data-quarto-table-cell-role="th">Rape</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Alabama</td>
<td>13.2</td>
<td>236</td>
<td>58</td>
<td>21.2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Alaska</td>
<td>10.0</td>
<td>263</td>
<td>48</td>
<td>44.5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Arizona</td>
<td>8.1</td>
<td>294</td>
<td>80</td>
<td>31.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Arkansas</td>
<td>8.8</td>
<td>190</td>
<td>50</td>
<td>19.5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">California</td>
<td>9.0</td>
<td>276</td>
<td>91</td>
<td>40.6</td>
</tr>
</tbody>
</table>


</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-96937f24-dc9f-4ad9-a08b-eb8dc0f3e511')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-96937f24-dc9f-4ad9-a08b-eb8dc0f3e511 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-96937f24-dc9f-4ad9-a08b-eb8dc0f3e511');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:13,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328496,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="dd00ac45-7917-403a-8ebe-5b2a6eece7f8" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1"></a>USArrests.columns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>Index(['Murder', 'Assault', 'UrbanPop', 'Rape'], dtype='object')</code></pre>
</div>
</div>
<p>This data contains states of US - Name - name of the states - Murder - number of arrests for murder per million pop. - Assault - number of arrests for assaults per million pop. - Urbanpop - percentage of population living in urban area - Rape - number of rape cases per million population</p>
<p>We first briefly examine the data. We notice that the variables have vastly different means.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:11,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328496,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="dcf61dec-5b76-4f6b-984a-9a1d1a5848bc" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1"></a>USArrests.mean()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>Murder        7.788
Assault     170.760
UrbanPop     65.540
Rape         21.232
dtype: float64</code></pre>
</div>
</div>
<p>We see that there are on average three times as many rapes as murders, and more than eight times as many assaults as rapes. We can also examine the variances of the four variables.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:9,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328496,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="9e3cb5bb-d661-4c15-94e2-6d4c2858fa99" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1"></a>USArrests.var()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<pre><code>Murder        18.970465
Assault     6945.165714
UrbanPop     209.518776
Rape          87.729159
dtype: float64</code></pre>
</div>
</div>
<p>Not surprisingly, the variables also have vastly different variances: the <code>UrbanPop</code> variable measures the percentage of the population in each state living in an urban area, which is not a comparable number to the number of rapes in each state per 100,000 individuals. If we failed to scale the variables before performing PCA, then most of the principal components that we observed would be driven by the <code>Assault</code> variable, since it has by far the largest mean and variance. Thus, it is important to standardize the variables to have mean zero and standard deviation one before performing PCA.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328497,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1"></a><span class="co"># scaling the data </span></span>
<span id="cb15-2"><a href="#cb15-2"></a>scaler <span class="op">=</span> StandardScaler()</span>
<span id="cb15-3"><a href="#cb15-3"></a>X_scaled <span class="op">=</span> scaler.fit_transform(USArrests)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>mean_</code> and <code>var_</code> components correspond to the means and variance of the variables that were used for scaling prior to implementing PCA.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328497,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="a358d41b-7d51-47ae-b14a-96d758cc3755" data-execution_count="11">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1"></a>scaler.mean_, scaler.var_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<pre><code>(array([  7.788, 170.76 ,  65.54 ,  21.232]),
 array([  18.591056, 6806.2624  ,  205.3284  ,   85.974576]))</code></pre>
</div>
</div>
<p>Now, the mean is approximately 0, and std is approximately 1 for all the columns</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:6,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328497,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="c96cea14-7e82-4b66-8bbe-a14e0713009b" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1"></a>X_scaled.mean(axis<span class="op">=</span><span class="dv">0</span>), X_scaled.var(axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>(array([-7.10542736e-17,  1.38777878e-16, -4.39648318e-16,  8.59312621e-16]),
 array([1., 1., 1., 1.]))</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:403,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328895,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="62c0f834-e775-4c22-e63f-1fd260379b6a" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1"></a><span class="co"># using pca to get principal components </span></span>
<span id="cb20-2"><a href="#cb20-2"></a>pca <span class="op">=</span> PCA()</span>
<span id="cb20-3"><a href="#cb20-3"></a>pca.fit(X_scaled)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>PCA()</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:9,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328895,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="5beb47ed-0ee6-4588-b0ef-76defa9062fc" data-execution_count="14">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1"></a><span class="co">#loading vectors</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>pd.DataFrame(pca.components_,columns <span class="op">=</span> USArrests.columns,index <span class="op">=</span> [<span class="st">'PC1'</span>,<span class="st">'PC2'</span>,<span class="st">'PC3'</span>,<span class="st">'PC4'</span>]).T</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">

  <div id="df-7969d8e6-fb39-4f00-a5b8-68f583753c33">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">PC1</th>
<th data-quarto-table-cell-role="th">PC2</th>
<th data-quarto-table-cell-role="th">PC3</th>
<th data-quarto-table-cell-role="th">PC4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Murder</td>
<td>0.535899</td>
<td>0.418181</td>
<td>-0.341233</td>
<td>0.649228</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Assault</td>
<td>0.583184</td>
<td>0.187986</td>
<td>-0.268148</td>
<td>-0.743407</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">UrbanPop</td>
<td>0.278191</td>
<td>-0.872806</td>
<td>-0.378016</td>
<td>0.133878</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Rape</td>
<td>0.543432</td>
<td>-0.167319</td>
<td>0.817778</td>
<td>0.089024</td>
</tr>
</tbody>
</table>


</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-7969d8e6-fb39-4f00-a5b8-68f583753c33')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-7969d8e6-fb39-4f00-a5b8-68f583753c33 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-7969d8e6-fb39-4f00-a5b8-68f583753c33');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<p>We see that there are four distinct principal components. This is to be expected because there are in general <span class="math inline">\(\min(n-1,p)\)</span> informative principal components in a data set with <span class="math inline">\(n\)</span> observations and <span class="math inline">\(p\)</span> variables.</p>
<p>Using the <code>PCA()</code> function, we do not need to explicitly multiply the data by the principal component loading vectors in order to obtain the principal component score vectors. Rather the <span class="math inline">\(50 \times 4\)</span> matrix <code>x</code> has as its columns the principal component score vectors. That is, the <span class="math inline">\(k\)</span>th column is the <span class="math inline">\(k\)</span>th principal component score vector.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901328895,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="4eeef67c-37ef-4e84-b99c-88e5b253f8ee" data-execution_count="15">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1"></a><span class="co"># getting the points in principal components</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>X_reduced <span class="op">=</span> pca.transform(X_scaled)</span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="bu">print</span>(X_reduced.shape)</span>
<span id="cb23-4"><a href="#cb23-4"></a>df <span class="op">=</span> pd.DataFrame(X_reduced ,columns <span class="op">=</span> [<span class="st">'PC1'</span>,<span class="st">'PC2'</span>,<span class="st">'PC3'</span>,<span class="st">'PC4'</span>])</span>
<span id="cb23-5"><a href="#cb23-5"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(50, 4)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="15">

  <div id="df-7c8cc65a-8def-4a24-93e2-186266b8009b">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">PC1</th>
<th data-quarto-table-cell-role="th">PC2</th>
<th data-quarto-table-cell-role="th">PC3</th>
<th data-quarto-table-cell-role="th">PC4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.985566</td>
<td>1.133392</td>
<td>-0.444269</td>
<td>0.156267</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1.950138</td>
<td>1.073213</td>
<td>2.040003</td>
<td>-0.438583</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>1.763164</td>
<td>-0.745957</td>
<td>0.054781</td>
<td>-0.834653</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>-0.141420</td>
<td>1.119797</td>
<td>0.114574</td>
<td>-0.182811</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>2.523980</td>
<td>-1.542934</td>
<td>0.598557</td>
<td>-0.341996</td>
</tr>
</tbody>
</table>


</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-7c8cc65a-8def-4a24-93e2-186266b8009b')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-7c8cc65a-8def-4a24-93e2-186266b8009b button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-7c8cc65a-8def-4a24-93e2-186266b8009b');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3433,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901332323,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="7eed813e-a11a-4bb6-c28f-85b39a8e0180" data-execution_count="16">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1"></a><span class="co"># biplot </span></span>
<span id="cb25-2"><a href="#cb25-2"></a>xvector <span class="op">=</span> pca.components_[<span class="dv">0</span>] </span>
<span id="cb25-3"><a href="#cb25-3"></a>yvector <span class="op">=</span> <span class="op">-</span>pca.components_[<span class="dv">1</span>]</span>
<span id="cb25-4"><a href="#cb25-4"></a></span>
<span id="cb25-5"><a href="#cb25-5"></a>xs <span class="op">=</span> X_reduced[:,<span class="dv">0</span>] </span>
<span id="cb25-6"><a href="#cb25-6"></a>ys <span class="op">=</span> <span class="op">-</span>X_reduced[:,<span class="dv">1</span>]</span>
<span id="cb25-7"><a href="#cb25-7"></a></span>
<span id="cb25-8"><a href="#cb25-8"></a>fig, ax1 <span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">1</span>, ncols<span class="op">=</span><span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">9</span>))</span>
<span id="cb25-9"><a href="#cb25-9"></a></span>
<span id="cb25-10"><a href="#cb25-10"></a><span class="co"># First Plot : Principal Component Scores 'DF_USArrests_PCA_scores" </span></span>
<span id="cb25-11"><a href="#cb25-11"></a>x_min, x_max <span class="op">=</span> xs.<span class="bu">min</span>() <span class="op">-</span> <span class="fl">0.5</span>, xs.<span class="bu">max</span>() <span class="op">+</span> <span class="fl">0.5</span></span>
<span id="cb25-12"><a href="#cb25-12"></a>y_min, y_max <span class="op">=</span> xs.<span class="bu">min</span>() <span class="op">-</span> <span class="fl">0.5</span>, xs.<span class="bu">max</span>() <span class="op">+</span> <span class="fl">0.5</span></span>
<span id="cb25-13"><a href="#cb25-13"></a>ax1.set_xlim(x_min, x_max)</span>
<span id="cb25-14"><a href="#cb25-14"></a>ax1.set_ylim(y_min, y_max)</span>
<span id="cb25-15"><a href="#cb25-15"></a></span>
<span id="cb25-16"><a href="#cb25-16"></a></span>
<span id="cb25-17"><a href="#cb25-17"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(xs)):</span>
<span id="cb25-18"><a href="#cb25-18"></a>    <span class="co"># circles project documents (ie rows from csv) as points onto PC axes</span></span>
<span id="cb25-19"><a href="#cb25-19"></a>    <span class="co"># plt.plot(xs[i], ys[i], 'bo')</span></span>
<span id="cb25-20"><a href="#cb25-20"></a>    <span class="co">#plt.text(xs[i], ys[i], USArrests.index[i], color='black', alpha=0.8, size=10)</span></span>
<span id="cb25-21"><a href="#cb25-21"></a>    ax1.annotate(USArrests.index[i], (xs[i], ys[i]), ha<span class="op">=</span><span class="st">'center'</span>, size<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb25-22"><a href="#cb25-22"></a></span>
<span id="cb25-23"><a href="#cb25-23"></a>ax1.set_xlabel(<span class="st">"1'st Principal Component Scores"</span>)</span>
<span id="cb25-24"><a href="#cb25-24"></a>ax1.set_ylabel(<span class="st">"2'nd Principal Component Scores"</span>)</span>
<span id="cb25-25"><a href="#cb25-25"></a></span>
<span id="cb25-26"><a href="#cb25-26"></a><span class="co"># Plot reference lines</span></span>
<span id="cb25-27"><a href="#cb25-27"></a>ax1.hlines(<span class="dv">0</span>,<span class="op">-</span><span class="fl">3.5</span>,<span class="fl">3.5</span>, linestyles<span class="op">=</span><span class="st">'dotted'</span>, colors<span class="op">=</span><span class="st">'grey'</span>)</span>
<span id="cb25-28"><a href="#cb25-28"></a>ax1.vlines(<span class="dv">0</span>,<span class="op">-</span><span class="fl">3.5</span>,<span class="fl">3.5</span>, linestyles<span class="op">=</span><span class="st">'dotted'</span>, colors<span class="op">=</span><span class="st">'grey'</span>)</span>
<span id="cb25-29"><a href="#cb25-29"></a></span>
<span id="cb25-30"><a href="#cb25-30"></a><span class="co"># Second Plot : Principal Component Loadings 'PCs' </span></span>
<span id="cb25-31"><a href="#cb25-31"></a>x_min, x_max <span class="op">=</span> <span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>          </span>
<span id="cb25-32"><a href="#cb25-32"></a>y_min, y_max <span class="op">=</span> <span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span></span>
<span id="cb25-33"><a href="#cb25-33"></a></span>
<span id="cb25-34"><a href="#cb25-34"></a>ax2 <span class="op">=</span> ax1.twinx().twiny()</span>
<span id="cb25-35"><a href="#cb25-35"></a>ax2.set_xlim(x_min, x_max)</span>
<span id="cb25-36"><a href="#cb25-36"></a>ax2.set_ylim(y_min, y_max)</span>
<span id="cb25-37"><a href="#cb25-37"></a></span>
<span id="cb25-38"><a href="#cb25-38"></a></span>
<span id="cb25-39"><a href="#cb25-39"></a></span>
<span id="cb25-40"><a href="#cb25-40"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(xvector)):</span>
<span id="cb25-41"><a href="#cb25-41"></a>    <span class="co"># arrows project features (ie columns from csv) as vectors onto PC axes</span></span>
<span id="cb25-42"><a href="#cb25-42"></a>    <span class="co">#plt.arrow(0, 0, xvector[i]*max(xs), yvector[i]*max(ys),</span></span>
<span id="cb25-43"><a href="#cb25-43"></a>    <span class="co">#          color='r', width=0.005, head_width=0.1)</span></span>
<span id="cb25-44"><a href="#cb25-44"></a>    <span class="co">#plt.text(xvector[i]*max(xs)*1.1, yvector[i]*max(ys)*1.1,</span></span>
<span id="cb25-45"><a href="#cb25-45"></a>    <span class="co">#         USArrests.columns[i], color='r', size=14)</span></span>
<span id="cb25-46"><a href="#cb25-46"></a>    ax2.arrow(<span class="dv">0</span>, <span class="dv">0</span>, xvector[i],  yvector[i], color<span class="op">=</span><span class="st">'red'</span>, width<span class="op">=</span><span class="fl">0.005</span>, head_width<span class="op">=</span><span class="fl">0.02</span>)</span>
<span id="cb25-47"><a href="#cb25-47"></a>    ax2.annotate(USArrests.columns[i], (xvector[i]<span class="op">*</span><span class="fl">1.05</span>, yvector[i]<span class="op">*</span><span class="fl">1.05</span>), color<span class="op">=</span><span class="st">'red'</span>, size<span class="op">=</span><span class="dv">14</span>)  </span>
<span id="cb25-48"><a href="#cb25-48"></a></span>
<span id="cb25-49"><a href="#cb25-49"></a>    </span>
<span id="cb25-50"><a href="#cb25-50"></a>ax2.set_xlabel(<span class="st">"1'st Principal Component (Loading)"</span>, color<span class="op">=</span><span class="st">'red'</span>)</span>
<span id="cb25-51"><a href="#cb25-51"></a>ax2.set_ylabel(<span class="st">"2'nd Principal Component (Loading)"</span>, color<span class="op">=</span><span class="st">'red'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>Text(0, 0.5, "2'nd Principal Component (Loading)")</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-16-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>The <code>PCA</code> function also outputs the variance of each principal component. For instance, on the <code>USArrests</code> data set, we can access these standard deviations as follows:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:10,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901332323,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="64bfd72a-eb84-4d39-81e1-cda0a4be5029" data-execution_count="17">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># varianc explained by each pc's</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>pca.explained_variance_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>array([2.53085875, 1.00996444, 0.36383998, 0.17696948])</code></pre>
</div>
</div>
<p>To get the proportion of variance explained by each principal component, we simply use the following attribute:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:9,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901332324,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="49b83cbe-7ce5-424e-b933-1de0062bc84b" data-execution_count="18">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1"></a>pca.explained_variance_<span class="op">/</span>np.var(X_reduced, ddof<span class="op">=</span><span class="dv">1</span>, axis<span class="op">=</span><span class="dv">0</span>).<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<pre><code>array([0.62006039, 0.24744129, 0.0891408 , 0.04335752])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:7,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901332324,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="8829cef4-3b0c-470e-bf2b-3190489c7acf" data-execution_count="19">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># variance explained ratio's </span></span>
<span id="cb31-2"><a href="#cb31-2"></a>pca.explained_variance_ratio_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>array([0.62006039, 0.24744129, 0.0891408 , 0.04335752])</code></pre>
</div>
</div>
<p>So, the first pc explaines 2.53 units of variance, which is around 62% of the total variance, the second pc expains 24.7 % of the variance in the data. We can plot the PVE explained by each component, as well as the cumulative PVE, as follows:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:883,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333201,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="1e955280-d60a-4050-a423-b8d6247288d4" data-execution_count="20">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1"></a><span class="co"># Percentage of Variance Explained (PVE) by each PC:</span></span>
<span id="cb33-2"><a href="#cb33-2"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb33-3"><a href="#cb33-3"></a>plt.ylim(<span class="dv">0</span>, <span class="fl">1.05</span>)</span>
<span id="cb33-4"><a href="#cb33-4"></a>index <span class="op">=</span> np.arange(pca.components_.shape[<span class="dv">0</span>])</span>
<span id="cb33-5"><a href="#cb33-5"></a>plt.bar(index, height<span class="op">=</span>pca.explained_variance_ratio_ , width<span class="op">=</span><span class="fl">0.3</span>, align<span class="op">=</span><span class="st">'center'</span>, </span>
<span id="cb33-6"><a href="#cb33-6"></a>        label<span class="op">=</span><span class="st">'percentage of variance explained'</span>)</span>
<span id="cb33-7"><a href="#cb33-7"></a>plt.plot(np.cumsum(pca.explained_variance_ratio_), <span class="st">'bo-'</span>, label<span class="op">=</span><span class="st">'cumulative PVE'</span>)</span>
<span id="cb33-8"><a href="#cb33-8"></a></span>
<span id="cb33-9"><a href="#cb33-9"></a>plt.ylabel(<span class="st">'percentage of variance explained'</span>)</span>
<span id="cb33-10"><a href="#cb33-10"></a>plt.xlabel(<span class="st">'Principal components'</span>)</span>
<span id="cb33-11"><a href="#cb33-11"></a>plt.legend(loc<span class="op">=</span><span class="st">'best'</span>)</span>
<span id="cb33-12"><a href="#cb33-12"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-20-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>You can refer to <a href="https://scikit-learn.org/stable/modules/decomposition.html#pca">https://scikit-learn.org/stable/modules/decomposition.html#pca</a> for more details</p>
</section>
<section id="matrix-completion" class="level3" data-number="10.1.2">
<h3 data-number="10.1.2" class="anchored" data-anchor-id="matrix-completion"><span class="header-section-number">10.1.2</span> Matrix Completion</h3>
<p>We now re-create the analysis carried out on the <code>USArrests</code> data in Section 12.3. We turn the data frame into a matrix, after centering and scaling each column to have mean zero and variance one.</p>
<p>We saw in Section 12.2.2 that solving the optimization problem~(12.6) on a centered data matrix <span class="math inline">\(\bf X\)</span> is equivalent to computing the first <span class="math inline">\(M\)</span> principal components of the data. The (SVD) is a general algorithm for solving (12.6).</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:7,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333201,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1"></a>u, d, v <span class="op">=</span> linalg.svd(X_scaled, full_matrices<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>svd()</code> function returns three components, <code>u</code>, <code>d</code>, and <code>v</code>. The matrix <code>v</code> is equivalent to the loading matrix from principal components (up to an unimportant sign flip).</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:7,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333202,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="7191d663-8815-4ced-ad51-83b26da39f30" data-execution_count="22">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1"></a>v.T</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>array([[-0.53589947,  0.41818087, -0.34123273,  0.6492278 ],
       [-0.58318363,  0.1879856 , -0.26814843, -0.74340748],
       [-0.27819087, -0.87280619, -0.37801579,  0.13387773],
       [-0.54343209, -0.16731864,  0.81777791,  0.08902432]])</code></pre>
</div>
</div>
<p>The matrix <code>u</code> is equivalent to the matrix of <em>standardized</em> scores, and the standard deviations are in the vector <code>d</code>. We can recover the score vectors using the output of <code>svd()</code>. They are identical to the score vectors output by <code>PCA()</code>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:5,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333202,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="a4363c48-fe9b-4062-c97c-bb79ee01dca5" data-execution_count="23">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1"></a>m, n <span class="op">=</span> X_scaled.shape</span>
<span id="cb37-2"><a href="#cb37-2"></a>u[:,:n].dot(np.diag(d))</span>
<span id="cb37-3"><a href="#cb37-3"></a>X_reduced</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>array([[ 9.85565885e-01,  1.13339238e+00, -4.44268788e-01,
         1.56267145e-01],
       [ 1.95013775e+00,  1.07321326e+00,  2.04000333e+00,
        -4.38583440e-01],
       [ 1.76316354e+00, -7.45956781e-01,  5.47808243e-02,
        -8.34652924e-01],
       [-1.41420290e-01,  1.11979678e+00,  1.14573692e-01,
        -1.82810896e-01],
       [ 2.52398013e+00, -1.54293399e+00,  5.98556799e-01,
        -3.41996478e-01],
       [ 1.51456286e+00, -9.87555085e-01,  1.09500699e+00,
         1.46488703e-03],
       [-1.35864746e+00, -1.08892789e+00, -6.43257569e-01,
        -1.18469414e-01],
       [ 4.77093091e-02, -3.25358925e-01, -7.18632942e-01,
        -8.81977637e-01],
       [ 3.01304227e+00,  3.92285133e-02, -5.76829492e-01,
        -9.62847520e-02],
       [ 1.63928304e+00,  1.27894240e+00, -3.42460080e-01,
         1.07679681e+00],
       [-9.12657146e-01, -1.57046001e+00,  5.07818939e-02,
         9.02806864e-01],
       [-1.63979985e+00,  2.10972917e-01,  2.59801342e-01,
        -4.99104101e-01],
       [ 1.37891072e+00, -6.81841189e-01, -6.77495641e-01,
        -1.22021292e-01],
       [-5.05461361e-01, -1.51562542e-01,  2.28054838e-01,
         4.24665700e-01],
       [-2.25364607e+00, -1.04054073e-01,  1.64564315e-01,
         1.75559157e-02],
       [-7.96881121e-01, -2.70164705e-01,  2.55533072e-02,
         2.06496428e-01],
       [-7.50859074e-01,  9.58440286e-01, -2.83694170e-02,
         6.70556671e-01],
       [ 1.56481798e+00,  8.71054655e-01, -7.83480358e-01,
         4.54728038e-01],
       [-2.39682949e+00,  3.76391576e-01, -6.56823864e-02,
        -3.30459817e-01],
       [ 1.76336939e+00,  4.27655192e-01, -1.57250127e-01,
        -5.59069521e-01],
       [-4.86166287e-01, -1.47449650e+00, -6.09497476e-01,
        -1.79598963e-01],
       [ 2.10844115e+00, -1.55396819e-01,  3.84868584e-01,
         1.02372019e-01],
       [-1.69268181e+00, -6.32261251e-01,  1.53070434e-01,
         6.73168850e-02],
       [ 9.96494459e-01,  2.39379599e+00, -7.40808395e-01,
         2.15508013e-01],
       [ 6.96787329e-01, -2.63354790e-01,  3.77443827e-01,
         2.25824461e-01],
       [-1.18545191e+00,  5.36874372e-01,  2.46889321e-01,
         1.23742227e-01],
       [-1.26563654e+00, -1.93953730e-01,  1.75573906e-01,
         1.58928878e-02],
       [ 2.87439454e+00, -7.75600196e-01,  1.16338049e+00,
         3.14515476e-01],
       [-2.38391541e+00, -1.80822897e-02,  3.68553932e-02,
        -3.31373376e-02],
       [ 1.81566110e-01, -1.44950571e+00, -7.64453551e-01,
         2.43382700e-01],
       [ 1.98002375e+00,  1.42848780e-01,  1.83692180e-01,
        -3.39533597e-01],
       [ 1.68257738e+00, -8.23184142e-01, -6.43075093e-01,
        -1.34843689e-02],
       [ 1.12337861e+00,  2.22800338e+00, -8.63571788e-01,
        -9.54381667e-01],
       [-2.99222562e+00,  5.99118824e-01,  3.01277285e-01,
        -2.53987327e-01],
       [-2.25965422e-01, -7.42238237e-01, -3.11391215e-02,
         4.73915911e-01],
       [-3.11782855e-01, -2.87854206e-01, -1.53097922e-02,
         1.03323208e-02],
       [ 5.91220768e-02, -5.41411454e-01,  9.39832977e-01,
        -2.37780688e-01],
       [-8.88415824e-01, -5.71100352e-01, -4.00628706e-01,
         3.59061124e-01],
       [-8.63772064e-01, -1.49197842e+00, -1.36994570e+00,
        -6.13569430e-01],
       [ 1.32072380e+00,  1.93340466e+00, -3.00537790e-01,
        -1.31466685e-01],
       [-1.98777484e+00,  8.23343241e-01,  3.89293329e-01,
        -1.09571764e-01],
       [ 9.99741684e-01,  8.60251305e-01,  1.88082949e-01,
         6.52864291e-01],
       [ 1.35513821e+00, -4.12480819e-01, -4.92068858e-01,
         6.43195491e-01],
       [-5.50565262e-01, -1.47150461e+00,  2.93728037e-01,
        -8.23140470e-02],
       [-2.80141174e+00,  1.40228806e+00,  8.41263094e-01,
        -1.44889914e-01],
       [-9.63349112e-02,  1.99735289e-01,  1.17125418e-02,
         2.11370813e-01],
       [-2.16903379e-01, -9.70124183e-01,  6.24870938e-01,
        -2.20847793e-01],
       [-2.10858541e+00,  1.42484670e+00,  1.04774671e-01,
         1.31908831e-01],
       [-2.07971417e+00, -6.11268624e-01, -1.38864998e-01,
         1.84103743e-01],
       [-6.29426664e-01,  3.21012967e-01, -2.40659234e-01,
        -1.66651801e-01]])</code></pre>
</div>
</div>
<p>We now omit 20 entries in the <span class="math inline">\(50\times 4\)</span> data matrix at random. We do so by first selecting 20 rows (states) at random, and then selecting one of the four entries in each row at random. This ensures that every row has at least three observed values.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:18,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333687,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="24">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1"></a>nomit <span class="op">=</span> <span class="dv">20</span></span>
<span id="cb39-2"><a href="#cb39-2"></a>np.random.seed(<span class="dv">2</span>)</span>
<span id="cb39-3"><a href="#cb39-3"></a>ina <span class="op">=</span> resample(<span class="bu">range</span>(<span class="dv">50</span>), n_samples <span class="op">=</span> nomit, replace<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb39-4"><a href="#cb39-4"></a>inb <span class="op">=</span> resample(<span class="bu">range</span>(<span class="dv">4</span>), n_samples <span class="op">=</span> nomit, replace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:18,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333687,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="9b3f0484-c2ec-4c94-ad27-c4fa49316651" data-execution_count="25">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1"></a>ina</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>[36, 47, 28, 9, 13, 0, 44, 46, 39, 23, 24, 48, 17, 12, 27, 33, 16, 2, 25, 14]</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:16,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333687,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1"></a>Xna <span class="op">=</span> X_scaled.copy()</span>
<span id="cb42-2"><a href="#cb42-2"></a>Xna[ina, inb] <span class="op">=</span> np.nan</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, <code>ina</code> contains 20 integers from 1 to 50; this represents the states that are selected to contain missing values. And <code>inb</code> contains 20 integers from 1 to 4, representing the features that contain the missing values for each of the selected states.</p>
<p>We now write some code to implement Algorithm 12.1. We first write a function that takes in a matrix, and returns an approximation to the matrix using the <code>svd()</code> function. This will be needed in Step 2 of Algorithm 12.1.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:16,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333687,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1"></a><span class="kw">def</span> SVD_approx(X, M<span class="op">=</span><span class="dv">1</span>):</span>
<span id="cb43-2"><a href="#cb43-2"></a>    u, d, v <span class="op">=</span> linalg.svd(X, full_matrices<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb43-3"><a href="#cb43-3"></a>    <span class="co">#m, n = X.shape</span></span>
<span id="cb43-4"><a href="#cb43-4"></a>    <span class="cf">return</span> u[:,:M].dot(np.diag(d[:M])).dot(v[:M,:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To conduct Step 1 of the algorithm, we initialize <code>Xhat</code>  this is <span class="math inline">\(\tilde{\bf X}\)</span> in Algorithm 12.1  by replacing the missing values with the column means of the non-missing entries.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:17,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333688,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="0f34772f-129b-40db-e532-9f35248fe9ab" data-execution_count="28">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1"></a>Xhat <span class="op">=</span> Xna.copy()</span>
<span id="cb44-2"><a href="#cb44-2"></a>xbar <span class="op">=</span> np.nanmean(Xna, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb44-3"><a href="#cb44-3"></a>xbar</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<pre><code>array([0.07184528, 0.02066806, 0.00883971, 0.01192302])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:14,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333688,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1"></a>Xhat[ina, inb] <span class="op">=</span> xbar[inb]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before we begin Step 2, we set ourselves up to measure the progress of our iterations:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:13,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333688,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="30">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1"></a>thresh <span class="op">=</span> <span class="fl">1e-7</span></span>
<span id="cb47-2"><a href="#cb47-2"></a>rel_err <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb47-3"><a href="#cb47-3"></a>itera <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb47-4"><a href="#cb47-4"></a>ismiss <span class="op">=</span> np.isnan(Xna)</span>
<span id="cb47-5"><a href="#cb47-5"></a>XX <span class="op">=</span> Xna<span class="op">-</span>xbar</span>
<span id="cb47-6"><a href="#cb47-6"></a>mssold <span class="op">=</span> np.mean((XX[<span class="op">~</span>ismiss])<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb47-7"><a href="#cb47-7"></a>mss0 <span class="op">=</span> np.mean(Xna[<span class="op">~</span>ismiss]<span class="op">**</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:14,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333689,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="1d808508-573f-4f7d-f4d1-3624ea3e16f3" data-execution_count="31">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1"></a>mssold, mss0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<pre><code>(1.0160715803147764, 1.0175207163820559)</code></pre>
</div>
</div>
<p>Here <code>ismiss</code> is a new logical matrix with the same dimensions as <code>Xna</code>; a given element equals <code>True</code> if the corresponding matrix element is missing. This is useful because it allows us to access both the missing and non-missing entries. We store the mean of the squared non-missing elements in <code>mss0</code>. We store the mean squared error of the non-missing elements of the old version of <code>Xhat</code> in <code>mssold</code>. We plan to store the mean squared error of the non-missing elements of the current version of <code>Xhat</code> in <code>mss</code>, and will then iterate Step 2 of Algorithm 12.1 until the <em>relative error</em>, defined as <code>(mssold - mss) / mss0</code>, falls below <code>thresh = 1e-7</code>. ( <em>Algorithm 12.1 tells us to iterate Step 2 until (12.14) is no longer decreasing. Determining whether (12.14) is decreasing requires us only to keep track of <code>mssold - mss</code>. However, in practice, we keep track of <code>(mssold - mss) / mss0</code> instead: this makes it so that the number of iterations required for Algorithm 12.1 to converge does not depend on whether we multiplied the raw data <span class="math inline">\(\bf X\)</span> by a constant factor.</em> )</p>
<p>In Step 2(a) of Algorithm 12.1, we approximate <code>Xhat</code> using <code>SVD_approx</code>; we call this <code>Xapp</code>. In Step 2(b), we use <code>Xapp</code> to update the estimates for elements in <code>Xhat</code> that are missing in <code>Xna</code>. Finally, in Step 2(c), we compute the relative error. These three steps are contained in this <code>while()</code> loop:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:12,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333689,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="b0e9c867-3b1c-4d58-979e-a87d3aec7a65" data-execution_count="32">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1"></a><span class="cf">while</span> rel_err <span class="op">&gt;</span> thresh:</span>
<span id="cb50-2"><a href="#cb50-2"></a>    itera <span class="op">=</span> itera <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb50-3"><a href="#cb50-3"></a>    <span class="co"># Step 2(a)</span></span>
<span id="cb50-4"><a href="#cb50-4"></a>    Xapp  <span class="op">=</span> SVD_approx(Xhat, M<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb50-5"><a href="#cb50-5"></a>    <span class="co"># Step 2(b)</span></span>
<span id="cb50-6"><a href="#cb50-6"></a>    Xhat[ismiss] <span class="op">=</span> Xapp[ismiss]</span>
<span id="cb50-7"><a href="#cb50-7"></a>    <span class="co"># Step 2(c)</span></span>
<span id="cb50-8"><a href="#cb50-8"></a>    XX <span class="op">=</span> Xna<span class="op">-</span>Xapp</span>
<span id="cb50-9"><a href="#cb50-9"></a>    mss <span class="op">=</span> np.mean((XX[<span class="op">~</span>ismiss])<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb50-10"><a href="#cb50-10"></a>    rel_err <span class="op">=</span> (mssold <span class="op">-</span> mss) <span class="op">/</span> mss0</span>
<span id="cb50-11"><a href="#cb50-11"></a>    mssold <span class="op">=</span> mss</span>
<span id="cb50-12"><a href="#cb50-12"></a>    <span class="bu">print</span>(<span class="st">"Iter: </span><span class="sc">%d</span><span class="st"> "</span><span class="op">%</span>itera)</span>
<span id="cb50-13"><a href="#cb50-13"></a>    <span class="bu">print</span>(<span class="st">"MSS: </span><span class="sc">%f</span><span class="st"> "</span><span class="op">%</span>mss)</span>
<span id="cb50-14"><a href="#cb50-14"></a>    <span class="bu">print</span>(<span class="st">"Rel. Err:</span><span class="sc">%f</span><span class="st"> </span><span class="ch">\n</span><span class="st">"</span><span class="op">%</span>rel_err)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Iter: 1 
MSS: 0.414600 
Rel. Err:0.591115 

Iter: 2 
MSS: 0.398214 
Rel. Err:0.016104 

Iter: 3 
MSS: 0.396375 
Rel. Err:0.001807 

Iter: 4 
MSS: 0.396105 
Rel. Err:0.000265 

Iter: 5 
MSS: 0.396051 
Rel. Err:0.000054 

Iter: 6 
MSS: 0.396037 
Rel. Err:0.000014 

Iter: 7 
MSS: 0.396033 
Rel. Err:0.000004 

Iter: 8 
MSS: 0.396032 
Rel. Err:0.000001 

Iter: 9 
MSS: 0.396032 
Rel. Err:0.000000 

Iter: 10 
MSS: 0.396031 
Rel. Err:0.000000 
</code></pre>
</div>
</div>
<p>We see that after eight iterations, the relative error has fallen below <code>thresh = 1e-7</code>, and so the algorithm terminates. When this happens, the mean squared error of the non-missing elements equals <span class="math inline">\(0.396\)</span>.</p>
<p>Finally, we compute the correlation between the 20 imputed values and the actual values:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:10,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333689,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="67eb4f20-271a-410d-92ba-bf1769546c9f" data-execution_count="33">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1"></a>np.corrcoef(Xhat[ismiss], X_scaled[ismiss])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="33">
<pre><code>array([[1.        , 0.80681102],
       [0.80681102, 1.        ]])</code></pre>
</div>
</div>
<p>In this lab, we implemented Algorithm 12.1 ourselves for didactic purposes. In practice, if you wishes to apply matrix completion to your data, you can use the <a href="https://github.com/iskandr/fancyimpute">https://github.com/iskandr/fancyimpute</a> package, which provides a very ecient implementation of a generalization of this algorithm.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:10,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333690,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="d5b08962-d440-47ff-eed0-5a26e1a2c378" data-execution_count="34">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1"></a>Xhat2 <span class="op">=</span> Xna.copy()</span>
<span id="cb54-2"><a href="#cb54-2"></a>X_filled <span class="op">=</span> SoftImpute(max_rank<span class="op">=</span><span class="dv">1</span>).fit_transform(Xhat2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[SoftImpute] Max Singular Value of X_init = 10.088074
[SoftImpute] Iter 1: observed MAE=0.477063 rank=1
[SoftImpute] Iter 2: observed MAE=0.463440 rank=1
[SoftImpute] Iter 3: observed MAE=0.460874 rank=1
[SoftImpute] Iter 4: observed MAE=0.460157 rank=1
[SoftImpute] Iter 5: observed MAE=0.459960 rank=1
[SoftImpute] Iter 6: observed MAE=0.459916 rank=1
[SoftImpute] Iter 7: observed MAE=0.459900 rank=1
[SoftImpute] Iter 8: observed MAE=0.459894 rank=1
[SoftImpute] Iter 9: observed MAE=0.459891 rank=1
[SoftImpute] Stopped after iteration 9 for lambda=0.201761</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901333690,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="64922dcc-2072-489d-a588-915208ba57cf" data-execution_count="35">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1"></a>np.corrcoef(X_filled[ismiss], X_scaled[ismiss])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="35">
<pre><code>array([[1.        , 0.80736346],
       [0.80736346, 1.        ]])</code></pre>
</div>
</div>
</section>
<section id="t-sne" class="level3" data-number="10.1.3">
<h3 data-number="10.1.3" class="anchored" data-anchor-id="t-sne"><span class="header-section-number">10.1.3</span> t-SNE</h3>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:380,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901334063,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="36">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1"></a>digits <span class="op">=</span> load_digits()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901334064,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="b4545f16-6617-42fa-f102-18614cb86a7e" data-execution_count="37">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1"></a>X <span class="op">=</span> digits.images.reshape(<span class="op">-</span><span class="dv">1</span>, digits.images.shape[<span class="dv">1</span>]<span class="op">*</span>digits.images.shape[<span class="dv">2</span>]) <span class="co">#vectorize</span></span>
<span id="cb59-2"><a href="#cb59-2"></a>X.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="37">
<pre><code>(1797, 64)</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2007,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901336068,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="e2dab58e-33e2-45c7-a0f7-494234e06260" data-execution_count="38">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1"></a>fig, ax_array <span class="op">=</span> plt.subplots(<span class="dv">5</span>, <span class="dv">5</span>)</span>
<span id="cb61-2"><a href="#cb61-2"></a>axes <span class="op">=</span> ax_array.flatten()</span>
<span id="cb61-3"><a href="#cb61-3"></a><span class="cf">for</span> i, ax <span class="kw">in</span> <span class="bu">enumerate</span>(axes):</span>
<span id="cb61-4"><a href="#cb61-4"></a>    ax.imshow(digits.images[i], cmap<span class="op">=</span><span class="st">'gray_r'</span>)</span>
<span id="cb61-5"><a href="#cb61-5"></a>plt.setp(axes, xticks<span class="op">=</span>[], yticks<span class="op">=</span>[], frame_on<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb61-6"><a href="#cb61-6"></a>plt.tight_layout(h_pad<span class="op">=</span><span class="fl">0.5</span>, w_pad<span class="op">=</span><span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-39-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901336069,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="39">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1"></a>tsne <span class="op">=</span> TSNE(</span>
<span id="cb62-2"><a href="#cb62-2"></a>    n_components<span class="op">=</span><span class="dv">2</span>, perplexity<span class="op">=</span><span class="dv">30</span>, init<span class="op">=</span><span class="st">"pca"</span>, learning_rate<span class="op">=</span><span class="st">"auto"</span>, random_state<span class="op">=</span><span class="dv">0</span>, verbose<span class="op">=</span><span class="va">True</span></span>
<span id="cb62-3"><a href="#cb62-3"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:34406,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901370472,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="b9b2595b-a3c9-476d-fe96-09458e5b2faf" data-execution_count="40">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1"></a>embedding <span class="op">=</span> tsne.fit_transform(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[t-SNE] Computing 91 nearest neighbors...
[t-SNE] Indexed 1797 samples in 0.001s...
[t-SNE] Computed neighbors for 1797 samples in 0.230s...
[t-SNE] Computed conditional probabilities for sample 1000 / 1797
[t-SNE] Computed conditional probabilities for sample 1797 / 1797
[t-SNE] Mean sigma: 11.585657</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.7/dist-packages/sklearn/manifold/_t_sne.py:986: FutureWarning: The PCA initialization in TSNE will change to have the standard deviation of PC1 equal to 1e-4 in 1.2. This will ensure better convergence.
  FutureWarning,</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[t-SNE] KL divergence after 250 iterations with early exaggeration: 54.981926
[t-SNE] KL divergence after 1000 iterations: 0.780030</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:796,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901371265,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="e75aa359-e710-4fd5-bb61-406348de05d6" data-execution_count="41">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb67-2"><a href="#cb67-2"></a>plt.scatter(embedding[:, <span class="dv">0</span>], embedding[:, <span class="dv">1</span>], c<span class="op">=</span>digits.target, cmap<span class="op">=</span><span class="st">'Spectral'</span>, s<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb67-3"><a href="#cb67-3"></a>plt.gca().set_aspect(<span class="st">'equal'</span>, <span class="st">'datalim'</span>)</span>
<span id="cb67-4"><a href="#cb67-4"></a>plt.colorbar(boundaries<span class="op">=</span>np.arange(<span class="dv">11</span>)<span class="op">-</span><span class="fl">0.5</span>).set_ticks(np.arange(<span class="dv">10</span>))</span>
<span id="cb67-5"><a href="#cb67-5"></a>plt.title(<span class="st">'tSNE of the Digits dataset'</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<pre><code>Text(0.5, 1.0, 'tSNE of the Digits dataset')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-42-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>openTSNE provides fast implementation and more optimization related to t-SNE, see <a href="https://opentsne.readthedocs.io/en/latest/">https://opentsne.readthedocs.io/en/latest/</a> for more details.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901371266,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="42">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1"></a>otsne <span class="op">=</span> oTSNE(</span>
<span id="cb69-2"><a href="#cb69-2"></a>    n_components<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb69-3"><a href="#cb69-3"></a>    perplexity<span class="op">=</span><span class="dv">30</span>,</span>
<span id="cb69-4"><a href="#cb69-4"></a>    initialization<span class="op">=</span><span class="st">'pca'</span>, </span>
<span id="cb69-5"><a href="#cb69-5"></a>    n_jobs<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb69-6"><a href="#cb69-6"></a>    random_state<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb69-7"><a href="#cb69-7"></a>    negative_gradient_method<span class="op">=</span><span class="st">'auto'</span>, </span>
<span id="cb69-8"><a href="#cb69-8"></a>    verbose<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb69-9"><a href="#cb69-9"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:15631,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901386894,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="a8fe3bf6-fc5a-4b58-c30d-23b52c6b20a1" data-execution_count="43">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1"></a>embedding <span class="op">=</span> otsne.fit(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------------------------------------------------------------------
TSNE(n_jobs=2, random_state=0, verbose=True)
--------------------------------------------------------------------------------
===&gt; Finding 90 nearest neighbors using Annoy approximate search using euclidean distance...
   --&gt; Time elapsed: 1.32 seconds
===&gt; Calculating affinity matrix...
   --&gt; Time elapsed: 0.29 seconds
===&gt; Calculating PCA-based initialization...
   --&gt; Time elapsed: 0.02 seconds
===&gt; Running optimization with exaggeration=12.00, lr=200.00 for 250 iterations...
Iteration   50, KL divergence 3.1044, 50 iterations in 0.7805 sec
Iteration  100, KL divergence 2.7275, 50 iterations in 0.4949 sec
Iteration  150, KL divergence 2.6611, 50 iterations in 0.6111 sec
Iteration  200, KL divergence 2.6358, 50 iterations in 0.5653 sec
Iteration  250, KL divergence 2.6226, 50 iterations in 0.7305 sec
   --&gt; Time elapsed: 3.19 seconds
===&gt; Running optimization with exaggeration=1.00, lr=200.00 for 500 iterations...
Iteration   50, KL divergence 1.2822, 50 iterations in 0.7395 sec
Iteration  100, KL divergence 1.0153, 50 iterations in 1.1167 sec
Iteration  150, KL divergence 0.9093, 50 iterations in 1.7734 sec
Iteration  200, KL divergence 0.8536, 50 iterations in 0.8590 sec
Iteration  250, KL divergence 0.8199, 50 iterations in 1.3595 sec
Iteration  300, KL divergence 0.7989, 50 iterations in 1.3390 sec
Iteration  350, KL divergence 0.7856, 50 iterations in 0.8678 sec
Iteration  400, KL divergence 0.7763, 50 iterations in 0.4503 sec
Iteration  450, KL divergence 0.7706, 50 iterations in 0.6760 sec
Iteration  500, KL divergence 0.7670, 50 iterations in 1.7542 sec
   --&gt; Time elapsed: 10.96 seconds</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1087,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901387978,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="47db771c-7940-4681-c8f8-0bcd0b7dfcc9" data-execution_count="44">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb72-2"><a href="#cb72-2"></a>plt.scatter(embedding[:, <span class="dv">0</span>], embedding[:, <span class="dv">1</span>], c<span class="op">=</span>digits.target, cmap<span class="op">=</span><span class="st">'Spectral'</span>, s<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb72-3"><a href="#cb72-3"></a>plt.gca().set_aspect(<span class="st">'equal'</span>, <span class="st">'datalim'</span>)</span>
<span id="cb72-4"><a href="#cb72-4"></a>plt.colorbar(boundaries<span class="op">=</span>np.arange(<span class="dv">11</span>)<span class="op">-</span><span class="fl">0.5</span>).set_ticks(np.arange(<span class="dv">10</span>))</span>
<span id="cb72-5"><a href="#cb72-5"></a>plt.title(<span class="st">'tSNE of the Digits dataset'</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="44">
<pre><code>Text(0.5, 1.0, 'tSNE of the Digits dataset')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-45-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="umap" class="level3" data-number="10.1.4">
<h3 data-number="10.1.4" class="anchored" data-anchor-id="umap"><span class="header-section-number">10.1.4</span> UMAP</h3>
<p>UMAP is useful for generating visualisations, but if you want to make use of UMAP more generally for machine learning tasks it is important to be be able to train a model and then later pass new data to the model and have it transform that data into the learned space. For example if we use UMAP to learn a latent space and then train a classifier on data transformed into the latent space then the classifier is only useful for prediction if we can transform data for which we want a prediction into the latent space the classifier uses.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:302,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901438208,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="47">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1"></a>X_train, X_test, y_train, y_test <span class="op">=</span> train_test_split(digits.data, digits.target, stratify<span class="op">=</span>digits.target, random_state<span class="op">=</span><span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now to get a benchmark idea of what we are looking at lets train a couple of different classifiers and then see how well they score on the test set. For this example lets try a support vector classifier and a KNN classifier.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:565,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901479182,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="3ae860e9-6dd2-46a5-896d-2466f5ac213b" data-execution_count="52">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1"></a>svc <span class="op">=</span> SVC(gamma<span class="op">=</span><span class="st">'auto'</span>).fit(X_train, y_train)</span>
<span id="cb75-2"><a href="#cb75-2"></a>knn <span class="op">=</span> KNeighborsClassifier().fit(X_train, y_train)</span>
<span id="cb75-3"><a href="#cb75-3"></a>svc.score(X_test, y_test), knn.score(X_test, y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="52">
<pre><code>(0.62, 0.9844444444444445)</code></pre>
</div>
</div>
<p>The goal now is to make use of UMAP as a preprocessing step that one could potentially fit into a pipeline.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:11524,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901493157,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="53">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1"></a>trans <span class="op">=</span> umap.UMAP(n_neighbors<span class="op">=</span><span class="dv">5</span>, random_state<span class="op">=</span><span class="dv">42</span>).fit(X_train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:957,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901494110,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="db9d9b67-48dc-40da-c3bd-8682a9bafa8f" data-execution_count="54">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb78-2"><a href="#cb78-2"></a>plt.scatter(trans.embedding_[:, <span class="dv">0</span>], trans.embedding_[:, <span class="dv">1</span>], c<span class="op">=</span>y_train, cmap<span class="op">=</span><span class="st">'Spectral'</span>, s<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb78-3"><a href="#cb78-3"></a>plt.gca().set_aspect(<span class="st">'equal'</span>, <span class="st">'datalim'</span>)</span>
<span id="cb78-4"><a href="#cb78-4"></a>plt.colorbar(boundaries<span class="op">=</span>np.arange(<span class="dv">11</span>)<span class="op">-</span><span class="fl">0.5</span>).set_ticks(np.arange(<span class="dv">10</span>))</span>
<span id="cb78-5"><a href="#cb78-5"></a>plt.title(<span class="st">'Umap of the Digits dataset'</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="54">
<pre><code>Text(0.5, 1.0, 'Umap of the Digits dataset')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-49-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>This looks very promising! Most of the classes got very cleanly separated, and that gives us some hope that it could help with classifier performance. We can now train some new models (again an SVC and a KNN classifier) on the embedded training data. This looks exactly as before but now we pass it the embedded data.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901494110,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="55">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1"></a>svc <span class="op">=</span> SVC(gamma<span class="op">=</span><span class="st">'auto'</span>).fit(trans.embedding_, y_train)</span>
<span id="cb80-2"><a href="#cb80-2"></a>knn <span class="op">=</span> KNeighborsClassifier().fit(trans.embedding_, y_train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8368,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901502476,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="56">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1"></a>test_embedding <span class="op">=</span> trans.transform(X_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The next important question is what the transform did to our test data. In principle we have a new two dimensional representation of the test-set, and ideally this should be based on the existing embedding of the training set</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:12,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901502477,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="6c5746b2-68e6-4746-dc4b-81a23f004a22" data-execution_count="57">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb82-2"><a href="#cb82-2"></a>plt.scatter(test_embedding[:, <span class="dv">0</span>], test_embedding[:, <span class="dv">1</span>], c<span class="op">=</span>y_test, cmap<span class="op">=</span><span class="st">'Spectral'</span>, s<span class="op">=</span><span class="dv">5</span>)</span>
<span id="cb82-3"><a href="#cb82-3"></a>plt.gca().set_aspect(<span class="st">'equal'</span>, <span class="st">'datalim'</span>)</span>
<span id="cb82-4"><a href="#cb82-4"></a>plt.colorbar(boundaries<span class="op">=</span>np.arange(<span class="dv">11</span>)<span class="op">-</span><span class="fl">0.5</span>).set_ticks(np.arange(<span class="dv">10</span>))</span>
<span id="cb82-5"><a href="#cb82-5"></a>plt.title(<span class="st">'Umap of the Digits dataset'</span>, fontsize<span class="op">=</span><span class="dv">24</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="57">
<pre><code>Text(0.5, 1.0, 'Umap of the Digits dataset')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-52-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>The results look like what we should expect; the test data has been embedded into two dimensions in exactly the locations we should expect (by class) given the embedding of the training data visualised above. This means we can now try out models that were trained on the embedded training data by handing them the newly transformed test set.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4840,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901507308,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="c2443467-f805-4d48-f745-5329285262cf" data-execution_count="58">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1"></a>svc.score(trans.transform(X_test), y_test), knn.score(trans.transform(X_test), y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="58">
<pre><code>(0.9822222222222222, 0.9822222222222222)</code></pre>
</div>
</div>
<p>The results are pretty good. While the accuracy of the KNN classifier did not improve there was not a lot of scope for improvement given the data. On the other hand the SVC has improved to have equal accuracy to the KNN classifier!</p>
<p>For more interesting datasets the larger dimensional embedding might have been a significant gain  it is certainly worth exploring as one of the parameters in a grid search across a pipeline that includes UMAP.</p>
</section>
</section>
<section id="clustering" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="clustering"><span class="header-section-number">10.2</span> Clustering</h2>
<section id="k-means-clustering" class="level3" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="k-means-clustering"><span class="header-section-number">10.2.1</span> <span class="math inline">\(K\)</span>-Means Clustering</h3>
<p>You can refer to <a href="https://scikit-learn.org/stable/modules/clustering.html">https://scikit-learn.org/stable/modules/clustering.html</a> for more details</p>
<p>The function <code>kmeans()</code> performs <span class="math inline">\(K\)</span>-means clustering in <code>Python</code>. We begin with a simple simulated example in which there truly are two clusters in the data: the first 25 observations have a mean shift relative to the next 25 observations.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:290,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901514121,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="59">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1"></a><span class="co"># set seed </span></span>
<span id="cb86-2"><a href="#cb86-2"></a>np.random.seed(<span class="dv">2</span>)</span>
<span id="cb86-3"><a href="#cb86-3"></a>X <span class="op">=</span> np.random.normal(size <span class="op">=</span> (<span class="dv">50</span>,<span class="dv">2</span>))</span>
<span id="cb86-4"><a href="#cb86-4"></a>X[:<span class="dv">25</span>,<span class="dv">0</span>] <span class="op">+=</span> <span class="dv">3</span> </span>
<span id="cb86-5"><a href="#cb86-5"></a>X[:<span class="dv">25</span>,<span class="dv">1</span>] <span class="op">-=</span> <span class="dv">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:354,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901514758,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="c4b481b3-d62e-40e2-c759-ae4e91a93a2a" data-execution_count="60">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1"></a>plt.scatter(X[:,<span class="dv">0</span>],X[:,<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="60">
<pre><code>&lt;matplotlib.collections.PathCollection at 0x7fdb94e66f50&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-55-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>We now perform <span class="math inline">\(K\)</span>-means clustering with <span class="math inline">\(K=2\)</span>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901516200,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="4b17d790-ed1d-4aae-cec6-f3da08c9cfec" data-execution_count="61">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1"></a>kmeans <span class="op">=</span> KMeans(n_clusters <span class="op">=</span> <span class="dv">2</span>, n_init<span class="op">=</span><span class="dv">20</span>, random_state<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb89-2"><a href="#cb89-2"></a>kmeans.fit(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="61">
<pre><code>KMeans(n_clusters=2, n_init=20, random_state=1)</code></pre>
</div>
</div>
<p>The cluster assignments of the 50 observations are listed as below:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901517043,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="12c254be-5400-4f1b-be38-b970a3110038" data-execution_count="62">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1"></a>kmeans.labels_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="62">
<pre><code>array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0], dtype=int32)</code></pre>
</div>
</div>
<p>The <span class="math inline">\(K\)</span>-means clustering nearly perfectly separated the observations into two clusters even though we did not supply any group information to <code>kmeans()</code>. We can plot the data, with each observation colored according to its cluster assignment.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901517670,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="4a543465-32f0-4861-900a-f38fe64e757f" data-execution_count="63">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1"></a>plt.scatter(X[:, <span class="dv">0</span>], X[:, <span class="dv">1</span>], c<span class="op">=</span>kmeans.labels_, s<span class="op">=</span><span class="dv">50</span>, cmap<span class="op">=</span>plt.cm.bwr)</span>
<span id="cb93-2"><a href="#cb93-2"></a></span>
<span id="cb93-3"><a href="#cb93-3"></a>centers <span class="op">=</span> kmeans.cluster_centers_</span>
<span id="cb93-4"><a href="#cb93-4"></a>plt.scatter(centers[:, <span class="dv">0</span>], centers[:, <span class="dv">1</span>],marker <span class="op">=</span> <span class="st">'*'</span>, c<span class="op">=</span><span class="st">'black'</span>, s<span class="op">=</span><span class="dv">200</span>, alpha<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb93-5"><a href="#cb93-5"></a>plt.xlabel(<span class="st">'X0'</span>)</span>
<span id="cb93-6"><a href="#cb93-6"></a>plt.ylabel(<span class="st">'X1'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="63">
<pre><code>Text(0, 0.5, 'X1')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-58-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901517994,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="aadf478e-0c46-4246-8b3c-8235bb28913d" data-execution_count="64">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1"></a>np.bincount(kmeans.labels_)</span>
<span id="cb95-2"><a href="#cb95-2"></a><span class="co"># 26 obs are classified in cluster 1, and 24 in cluster 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="64">
<pre><code>array([26, 24])</code></pre>
</div>
</div>
<p>Here the observations can be easily plotted because they are two-dimensional. If there were more than two variables then we could instead perform PCA and plot the first two principal components score vectors.</p>
<p>In this example, we knew that there really were two clusters because we generated the data. However, for real data, in general we do not know the true number of clusters. We could instead have performed <span class="math inline">\(K\)</span>-means clustering on this example with <span class="math inline">\(K=3\)</span>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:415,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901518793,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="89ca673e-82c0-4940-aa39-ea3bac899a79" data-execution_count="65">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1"></a>kmean2 <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">3</span>, n_init<span class="op">=</span><span class="dv">20</span>, random_state<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb97-2"><a href="#cb97-2"></a>kmean2.fit(X)</span>
<span id="cb97-3"><a href="#cb97-3"></a></span>
<span id="cb97-4"><a href="#cb97-4"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">6</span>,<span class="dv">5</span>))</span>
<span id="cb97-5"><a href="#cb97-5"></a>plt.scatter(X[:,<span class="dv">0</span>], X[:,<span class="dv">1</span>], s<span class="op">=</span><span class="dv">50</span>, c<span class="op">=</span>kmean2.labels_, cmap<span class="op">=</span>plt.cm.prism) </span>
<span id="cb97-6"><a href="#cb97-6"></a>plt.scatter(kmean2.cluster_centers_[:, <span class="dv">0</span>], kmean2.cluster_centers_[:, <span class="dv">1</span>], marker<span class="op">=</span><span class="st">'*'</span>, s<span class="op">=</span><span class="dv">150</span>,</span>
<span id="cb97-7"><a href="#cb97-7"></a>            color<span class="op">=</span><span class="st">'blue'</span>, label<span class="op">=</span><span class="st">'Centers'</span>)</span>
<span id="cb97-8"><a href="#cb97-8"></a>plt.legend(loc<span class="op">=</span><span class="st">'best'</span>)</span>
<span id="cb97-9"><a href="#cb97-9"></a>plt.xlabel(<span class="st">'X0'</span>)</span>
<span id="cb97-10"><a href="#cb97-10"></a>plt.ylabel(<span class="st">'X1'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="65">
<pre><code>Text(0, 0.5, 'X1')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-60-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>When <span class="math inline">\(K=3\)</span>, <span class="math inline">\(K\)</span>-means clustering splits up the two clusters.</p>
<p>To run the <code>kmeans()</code> function in <code>Python</code> with multiple initial cluster assignments, we use the <code>n_init</code> argument. If a value of <code>n_init</code> greater than one is used, then <span class="math inline">\(K\)</span>-means clustering will be performed using multiple random assignments in Step~1 of Algorithm 12.2, and the <code>kmeans()</code> function will report only the best results. Here we compare using <code>n_init = 1</code> to <code>n_init = 20</code>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:321,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901519112,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="987e6201-4e0d-4501-9a89-ce32d5903fc0" data-execution_count="66">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1"></a>kmeans3 <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">3</span>, n_init<span class="op">=</span><span class="dv">1</span>, random_state<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb99-2"><a href="#cb99-2"></a>kmeans3.fit(X)</span>
<span id="cb99-3"><a href="#cb99-3"></a>kmeans4 <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">3</span>,  n_init<span class="op">=</span><span class="dv">20</span>, random_state<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb99-4"><a href="#cb99-4"></a>kmeans4.fit(X)</span>
<span id="cb99-5"><a href="#cb99-5"></a></span>
<span id="cb99-6"><a href="#cb99-6"></a><span class="bu">print</span>(<span class="st">'inertia with n_init=1:'</span>, kmeans3.inertia_)</span>
<span id="cb99-7"><a href="#cb99-7"></a><span class="bu">print</span>(<span class="st">'inertia with n_init=20:'</span>, kmeans4.inertia_)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>inertia with n_init=1: 69.42547524486459
inertia with n_init=20: 68.97379200939723</code></pre>
</div>
</div>
<p>Note that <code>kmeans.inertia_</code> is the sum of squared distances of samples to their closest cluster center which we seek to minimize by performing <span class="math inline">\(K\)</span>-means clustering. We <em>strongly</em> recommend always running <span class="math inline">\(K\)</span>-means clustering with a large value of <code>n_init</code>, such as 20 or 50, since otherwise an undesirable local optimum may be obtained.</p>
<p>When performing <span class="math inline">\(K\)</span>-means clustering, in addition to using multiple initial cluster assignments, it is also important to set a random seed using the <code>random_state</code>. This way, the initial cluster assignments in Step~1 can be replicated, and the <span class="math inline">\(K\)</span>-means output will be fully reproducible.</p>
</section>
<section id="hierarchical-clustering" class="level3" data-number="10.2.2">
<h3 data-number="10.2.2" class="anchored" data-anchor-id="hierarchical-clustering"><span class="header-section-number">10.2.2</span> Hierarchical Clustering</h3>
<p>The <code>AgglomerativeClustering()</code> function implements hierarchical clustering in <code>Python</code>. In the following example we use the data from the previous lab to plot the hierarchical clustering dendrogram using complete, single, and average linkage clustering, with Euclidean distance as the dissimilarity measure. We begin by clustering observations using complete linkage.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901520181,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="09705ac4-2f65-4ebb-9e5f-4fa01331c0ba" data-execution_count="67">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1"></a>hclust <span class="op">=</span> AgglomerativeClustering(n_clusters<span class="op">=</span><span class="dv">2</span>, linkage<span class="op">=</span><span class="st">'complete'</span>, compute_distances<span class="op">=</span><span class="va">True</span>)    </span>
<span id="cb101-2"><a href="#cb101-2"></a>hclust.fit(X)</span>
<span id="cb101-3"><a href="#cb101-3"></a>hclust.labels_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="67">
<pre><code>array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])</code></pre>
</div>
</div>
<p>We could just as easily perform hierarchical clustering with average or single linkage instead:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901522508,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="a774e2b0-a3c8-46aa-92f3-8681f7ccff82" data-execution_count="68">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1"></a>hclust <span class="op">=</span> AgglomerativeClustering(n_clusters<span class="op">=</span><span class="dv">2</span>, linkage<span class="op">=</span><span class="st">'average'</span>, compute_distances<span class="op">=</span><span class="va">True</span>)    </span>
<span id="cb103-2"><a href="#cb103-2"></a>hclust.fit(X)</span>
<span id="cb103-3"><a href="#cb103-3"></a>hclust.labels_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="68">
<pre><code>array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901522970,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="bed1d5d9-8e4c-4e73-f5d4-7afa2d65596e" data-execution_count="69">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1"></a>hclust <span class="op">=</span> AgglomerativeClustering(n_clusters<span class="op">=</span><span class="dv">2</span>, linkage<span class="op">=</span><span class="st">'single'</span>, compute_distances<span class="op">=</span><span class="va">True</span>)    </span>
<span id="cb105-2"><a href="#cb105-2"></a>hclust.fit(X)</span>
<span id="cb105-3"><a href="#cb105-3"></a>hclust.labels_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="69">
<pre><code>array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])</code></pre>
</div>
</div>
<p>We can now plot the dendrograms obtained using the custom function. The numbers at the bottom of the plot identify each observation.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:960,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901543418,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="c343ea52-b29a-4ef0-e571-48c164aadb62" data-execution_count="70">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1"></a>plt.title(<span class="st">"Hierarchical Clustering Dendrogram"</span>)</span>
<span id="cb107-2"><a href="#cb107-2"></a><span class="co"># plot the top three levels of the dendrogram</span></span>
<span id="cb107-3"><a href="#cb107-3"></a>plot_dendrogram(hclust, truncate_mode<span class="op">=</span><span class="st">"level"</span>, p<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb107-4"><a href="#cb107-4"></a>plt.xlabel(<span class="st">"Number of points in node (or index of point if no parenthesis)."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="70">
<pre><code>Text(0.5, 0, 'Number of points in node (or index of point if no parenthesis).')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-65-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>You can also perform clustering using Scipy, see <a href="https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html">https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html</a> for more details.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901544563,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="71">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1"></a>hc_complete <span class="op">=</span> linkage(X, method<span class="op">=</span><span class="st">'complete'</span>)</span>
<span id="cb109-2"><a href="#cb109-2"></a>hc_average <span class="op">=</span> linkage(X, method<span class="op">=</span><span class="st">'average'</span>)</span>
<span id="cb109-3"><a href="#cb109-3"></a>hc_single <span class="op">=</span> linkage(X, method<span class="op">=</span><span class="st">'single'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:5763,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901550721,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="f283a7c9-d7f8-4baf-9b69-f5246a82f5ca" data-execution_count="72">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1"></a>f, axes <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">3</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb110-2"><a href="#cb110-2"></a></span>
<span id="cb110-3"><a href="#cb110-3"></a>f.set_figheight(<span class="dv">8</span>)</span>
<span id="cb110-4"><a href="#cb110-4"></a>f.set_figwidth(<span class="dv">16</span>)</span>
<span id="cb110-5"><a href="#cb110-5"></a></span>
<span id="cb110-6"><a href="#cb110-6"></a>dendrogram(hc_complete,</span>
<span id="cb110-7"><a href="#cb110-7"></a>           labels<span class="op">=</span>X,</span>
<span id="cb110-8"><a href="#cb110-8"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb110-9"><a href="#cb110-9"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb110-10"><a href="#cb110-10"></a>           ax<span class="op">=</span>axes[<span class="dv">0</span>])</span>
<span id="cb110-11"><a href="#cb110-11"></a></span>
<span id="cb110-12"><a href="#cb110-12"></a>dendrogram(hc_average,</span>
<span id="cb110-13"><a href="#cb110-13"></a>           labels<span class="op">=</span>X, <span class="co"># coordinates</span></span>
<span id="cb110-14"><a href="#cb110-14"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb110-15"><a href="#cb110-15"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb110-16"><a href="#cb110-16"></a>           ax<span class="op">=</span>axes[<span class="dv">1</span>])</span>
<span id="cb110-17"><a href="#cb110-17"></a></span>
<span id="cb110-18"><a href="#cb110-18"></a>dendrogram(hc_single,</span>
<span id="cb110-19"><a href="#cb110-19"></a>           labels<span class="op">=</span>X,</span>
<span id="cb110-20"><a href="#cb110-20"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb110-21"><a href="#cb110-21"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb110-22"><a href="#cb110-22"></a>           ax<span class="op">=</span>axes[<span class="dv">2</span>])</span>
<span id="cb110-23"><a href="#cb110-23"></a></span>
<span id="cb110-24"><a href="#cb110-24"></a>axes[<span class="dv">0</span>].set_title(<span class="st">'Complete Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb110-25"><a href="#cb110-25"></a>axes[<span class="dv">1</span>].set_title(<span class="st">'Average Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb110-26"><a href="#cb110-26"></a>axes[<span class="dv">2</span>].set_title(<span class="st">'Single Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.7/dist-packages/matplotlib/text.py:1165: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if s != self._text:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="72">
<pre><code>Text(0.5, 1.0, 'Single Linkage')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-67-output-3.png" class="img-fluid"></p>
</div>
</div>
<p>To determine the cluster labels for each observation associated with a given cut of the dendrogram, we can use the <code>cutree()</code> function: The second argument to <code>cutree()</code> is the number of clusters we wish to obtain. For this data, complete and average linkage generally separate the observations into their correct groups. However, single linkage does not.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:6,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901550721,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="945083e3-60c1-4e6d-9f1f-265deb4d3188" data-execution_count="73">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1"></a>cut_tree(hc_complete, n_clusters<span class="op">=</span><span class="dv">2</span>, height<span class="op">=</span><span class="va">None</span>).ravel()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="73">
<pre><code>array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901551464,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="6143f4f1-10fb-45ea-ca42-d336ac479ac4" data-execution_count="74">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1"></a>cut_tree(hc_average, <span class="dv">2</span>).ravel()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="74">
<pre><code>array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901551754,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="345297bc-0beb-4f1c-c81f-9b31795d783c" data-execution_count="75">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1"></a>cut_tree(hc_single, <span class="dv">2</span>).ravel()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="75">
<pre><code>array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1])</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:5,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901552575,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="1a70db11-88aa-4c7a-df74-85a7674e9ef3" data-execution_count="76">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1"></a>cut_tree(hc_single, <span class="dv">3</span>).ravel()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="76">
<pre><code>array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 2, 2, 2, 2, 1, 2, 2, 2, 1, 2, 2, 2, 2, 2, 1, 2, 2, 1, 2,
       2, 2, 2, 2, 2, 1])</code></pre>
</div>
</div>
<p>To scale the variables before performing hierarchical clustering of the observations:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1814,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901555191,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="b187e742-0426-4d36-c5e0-e24b2560114f" data-execution_count="77">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1"></a>scaler <span class="op">=</span> StandardScaler()</span>
<span id="cb121-2"><a href="#cb121-2"></a>X_scaled <span class="op">=</span> scaler.fit_transform(X)</span>
<span id="cb121-3"><a href="#cb121-3"></a></span>
<span id="cb121-4"><a href="#cb121-4"></a>dendrogram(linkage(X_scaled, method<span class="op">=</span><span class="st">'complete'</span>),</span>
<span id="cb121-5"><a href="#cb121-5"></a>           labels<span class="op">=</span>X_scaled,</span>
<span id="cb121-6"><a href="#cb121-6"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb121-7"><a href="#cb121-7"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb121-8"><a href="#cb121-8"></a></span>
<span id="cb121-9"><a href="#cb121-9"></a>plt.title(<span class="st">'Hierarchical Clustering with Scaled Features'</span>, size<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.7/dist-packages/matplotlib/text.py:1165: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if s != self._text:</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="77">
<pre><code>Text(0.5, 1.0, 'Hierarchical Clustering with Scaled Features')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-72-output-3.png" class="img-fluid"></p>
</div>
</div>
<p>Correlation-based distance can be computed using the <code>pdist()</code> function, which converts an arbitrary square symmetric matrix into a form that the <code>scipy</code> recognizes as a distance matrix. However, this only makes sense for data with at least three features since the absolute correlation between any two observations with measurements on two features is always 1. Hence, we will cluster a three-dimensional data set. This data set does not contain any true clusters.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1359,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901558355,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="598c275c-8d8a-4d26-b30b-843a3112818b" data-execution_count="78">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1"></a>X <span class="op">=</span> np.random.normal(size<span class="op">=</span><span class="dv">30</span><span class="op">*</span><span class="dv">3</span>).reshape(<span class="dv">30</span>, <span class="dv">3</span>)</span>
<span id="cb124-2"><a href="#cb124-2"></a></span>
<span id="cb124-3"><a href="#cb124-3"></a><span class="co"># scipy linkage takes care of the distance function pdist</span></span>
<span id="cb124-4"><a href="#cb124-4"></a>dendrogram(linkage(X, method<span class="op">=</span><span class="st">'complete'</span>, metric<span class="op">=</span><span class="st">'correlation'</span>),</span>
<span id="cb124-5"><a href="#cb124-5"></a>           labels<span class="op">=</span>X,</span>
<span id="cb124-6"><a href="#cb124-6"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb124-7"><a href="#cb124-7"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb124-8"><a href="#cb124-8"></a></span>
<span id="cb124-9"><a href="#cb124-9"></a>plt.title(<span class="st">'Complete Linkage with Correlation Based Distance'</span>, size<span class="op">=</span><span class="dv">16</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-73-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>You can get a more efficient implementation at <a href="http://danifold.net/fastcluster.html">http://danifold.net/fastcluster.html</a>. The usuage is the same as Scipy.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:811,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901559164,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="8b637978-3c07-4bf0-f4c5-de9c77b7b123" data-execution_count="79">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb125-1"><a href="#cb125-1"></a>lk <span class="op">=</span> fc.linkage(X, method<span class="op">=</span><span class="st">'complete'</span>, metric<span class="op">=</span><span class="st">'correlation'</span>, preserve_input<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb125-2"><a href="#cb125-2"></a>dendrogram(lk,</span>
<span id="cb125-3"><a href="#cb125-3"></a>           labels<span class="op">=</span>X,</span>
<span id="cb125-4"><a href="#cb125-4"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb125-5"><a href="#cb125-5"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb125-6"><a href="#cb125-6"></a></span>
<span id="cb125-7"><a href="#cb125-7"></a>plt.title(<span class="st">'Complete Linkage with Correlation Based Distance'</span>, size<span class="op">=</span><span class="dv">16</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-74-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="dbscan" class="level3" data-number="10.2.3">
<h3 data-number="10.2.3" class="anchored" data-anchor-id="dbscan"><span class="header-section-number">10.2.3</span> DBSCAN</h3>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:275,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901561034,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="80">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1"></a><span class="co"># Generate sample data</span></span>
<span id="cb126-2"><a href="#cb126-2"></a>centers <span class="op">=</span> [[<span class="dv">1</span>, <span class="dv">1</span>], [<span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>], [<span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb126-3"><a href="#cb126-3"></a>X, labels_true <span class="op">=</span> make_blobs(</span>
<span id="cb126-4"><a href="#cb126-4"></a>    n_samples<span class="op">=</span><span class="dv">750</span>, centers<span class="op">=</span>centers, cluster_std<span class="op">=</span><span class="fl">0.4</span>, random_state<span class="op">=</span><span class="dv">0</span></span>
<span id="cb126-5"><a href="#cb126-5"></a>)</span>
<span id="cb126-6"><a href="#cb126-6"></a></span>
<span id="cb126-7"><a href="#cb126-7"></a>X <span class="op">=</span> StandardScaler().fit_transform(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For more clustering performance measure see <a href="https://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation">https://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation</a>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901563199,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="2ddf0aed-5516-4421-c167-b4394e4a6f3a" data-execution_count="81">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1"></a><span class="co"># Compute DBSCAN</span></span>
<span id="cb127-2"><a href="#cb127-2"></a>db <span class="op">=</span> DBSCAN(eps<span class="op">=</span><span class="fl">0.3</span>, min_samples<span class="op">=</span><span class="dv">10</span>).fit(X)</span>
<span id="cb127-3"><a href="#cb127-3"></a>core_samples_mask <span class="op">=</span> np.zeros_like(db.labels_, dtype<span class="op">=</span><span class="bu">bool</span>)</span>
<span id="cb127-4"><a href="#cb127-4"></a>core_samples_mask[db.core_sample_indices_] <span class="op">=</span> <span class="va">True</span></span>
<span id="cb127-5"><a href="#cb127-5"></a>labels <span class="op">=</span> db.labels_</span>
<span id="cb127-6"><a href="#cb127-6"></a></span>
<span id="cb127-7"><a href="#cb127-7"></a><span class="co"># Number of clusters in labels, ignoring noise if present.</span></span>
<span id="cb127-8"><a href="#cb127-8"></a>n_clusters_ <span class="op">=</span> <span class="bu">len</span>(<span class="bu">set</span>(labels)) <span class="op">-</span> (<span class="dv">1</span> <span class="cf">if</span> <span class="op">-</span><span class="dv">1</span> <span class="kw">in</span> labels <span class="cf">else</span> <span class="dv">0</span>)</span>
<span id="cb127-9"><a href="#cb127-9"></a>n_noise_ <span class="op">=</span> <span class="bu">list</span>(labels).count(<span class="op">-</span><span class="dv">1</span>)</span>
<span id="cb127-10"><a href="#cb127-10"></a></span>
<span id="cb127-11"><a href="#cb127-11"></a><span class="bu">print</span>(<span class="st">"Estimated number of clusters: </span><span class="sc">%d</span><span class="st">"</span> <span class="op">%</span> n_clusters_)</span>
<span id="cb127-12"><a href="#cb127-12"></a><span class="bu">print</span>(<span class="st">"Estimated number of noise points: </span><span class="sc">%d</span><span class="st">"</span> <span class="op">%</span> n_noise_)</span>
<span id="cb127-13"><a href="#cb127-13"></a><span class="bu">print</span>(<span class="st">"Homogeneity: </span><span class="sc">%0.3f</span><span class="st">"</span> <span class="op">%</span> metrics.homogeneity_score(labels_true, labels))</span>
<span id="cb127-14"><a href="#cb127-14"></a><span class="bu">print</span>(<span class="st">"Completeness: </span><span class="sc">%0.3f</span><span class="st">"</span> <span class="op">%</span> metrics.completeness_score(labels_true, labels))</span>
<span id="cb127-15"><a href="#cb127-15"></a><span class="bu">print</span>(<span class="st">"V-measure: </span><span class="sc">%0.3f</span><span class="st">"</span> <span class="op">%</span> metrics.v_measure_score(labels_true, labels))</span>
<span id="cb127-16"><a href="#cb127-16"></a><span class="bu">print</span>(<span class="st">"Adjusted Rand Index: </span><span class="sc">%0.3f</span><span class="st">"</span> <span class="op">%</span> metrics.adjusted_rand_score(labels_true, labels))</span>
<span id="cb127-17"><a href="#cb127-17"></a><span class="bu">print</span>(</span>
<span id="cb127-18"><a href="#cb127-18"></a>    <span class="st">"Adjusted Mutual Information: </span><span class="sc">%0.3f</span><span class="st">"</span></span>
<span id="cb127-19"><a href="#cb127-19"></a>    <span class="op">%</span> metrics.adjusted_mutual_info_score(labels_true, labels)</span>
<span id="cb127-20"><a href="#cb127-20"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Estimated number of clusters: 3
Estimated number of noise points: 18
Homogeneity: 0.953
Completeness: 0.883
V-measure: 0.917
Adjusted Rand Index: 0.952
Adjusted Mutual Information: 0.916</code></pre>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:622,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901565302,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="f6f7d87c-cd78-4108-e196-7b432ef43b5d" data-execution_count="82">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1"></a><span class="co"># Black removed and is used for noise instead.</span></span>
<span id="cb129-2"><a href="#cb129-2"></a>unique_labels <span class="op">=</span> <span class="bu">set</span>(labels)</span>
<span id="cb129-3"><a href="#cb129-3"></a>colors <span class="op">=</span> [plt.cm.Spectral(each) <span class="cf">for</span> each <span class="kw">in</span> np.linspace(<span class="dv">0</span>, <span class="dv">1</span>, <span class="bu">len</span>(unique_labels))]</span>
<span id="cb129-4"><a href="#cb129-4"></a><span class="cf">for</span> k, col <span class="kw">in</span> <span class="bu">zip</span>(unique_labels, colors):</span>
<span id="cb129-5"><a href="#cb129-5"></a>    <span class="cf">if</span> k <span class="op">==</span> <span class="op">-</span><span class="dv">1</span>:</span>
<span id="cb129-6"><a href="#cb129-6"></a>        <span class="co"># Black used for noise.</span></span>
<span id="cb129-7"><a href="#cb129-7"></a>        col <span class="op">=</span> [<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>]</span>
<span id="cb129-8"><a href="#cb129-8"></a></span>
<span id="cb129-9"><a href="#cb129-9"></a>    class_member_mask <span class="op">=</span> labels <span class="op">==</span> k</span>
<span id="cb129-10"><a href="#cb129-10"></a></span>
<span id="cb129-11"><a href="#cb129-11"></a>    xy <span class="op">=</span> X[class_member_mask <span class="op">&amp;</span> core_samples_mask]</span>
<span id="cb129-12"><a href="#cb129-12"></a>    plt.plot(</span>
<span id="cb129-13"><a href="#cb129-13"></a>        xy[:, <span class="dv">0</span>],</span>
<span id="cb129-14"><a href="#cb129-14"></a>        xy[:, <span class="dv">1</span>],</span>
<span id="cb129-15"><a href="#cb129-15"></a>        <span class="st">"o"</span>,</span>
<span id="cb129-16"><a href="#cb129-16"></a>        markerfacecolor<span class="op">=</span><span class="bu">tuple</span>(col),</span>
<span id="cb129-17"><a href="#cb129-17"></a>        markeredgecolor<span class="op">=</span><span class="st">"k"</span>,</span>
<span id="cb129-18"><a href="#cb129-18"></a>        markersize<span class="op">=</span><span class="dv">14</span>,</span>
<span id="cb129-19"><a href="#cb129-19"></a>    )</span>
<span id="cb129-20"><a href="#cb129-20"></a></span>
<span id="cb129-21"><a href="#cb129-21"></a>    xy <span class="op">=</span> X[class_member_mask <span class="op">&amp;</span> <span class="op">~</span>core_samples_mask]</span>
<span id="cb129-22"><a href="#cb129-22"></a>    plt.plot(</span>
<span id="cb129-23"><a href="#cb129-23"></a>        xy[:, <span class="dv">0</span>],</span>
<span id="cb129-24"><a href="#cb129-24"></a>        xy[:, <span class="dv">1</span>],</span>
<span id="cb129-25"><a href="#cb129-25"></a>        <span class="st">"o"</span>,</span>
<span id="cb129-26"><a href="#cb129-26"></a>        markerfacecolor<span class="op">=</span><span class="bu">tuple</span>(col),</span>
<span id="cb129-27"><a href="#cb129-27"></a>        markeredgecolor<span class="op">=</span><span class="st">"k"</span>,</span>
<span id="cb129-28"><a href="#cb129-28"></a>        markersize<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb129-29"><a href="#cb129-29"></a>    )</span>
<span id="cb129-30"><a href="#cb129-30"></a></span>
<span id="cb129-31"><a href="#cb129-31"></a>plt.title(<span class="st">"Estimated number of clusters: </span><span class="sc">%d</span><span class="st">"</span> <span class="op">%</span> n_clusters_)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="82">
<pre><code>Text(0.5, 1.0, 'Estimated number of clusters: 3')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-77-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>The DBSCAN, however, is quite sensitive to the hypeparameter</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901567108,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="83">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1"></a><span class="co"># Generate sample data</span></span>
<span id="cb131-2"><a href="#cb131-2"></a>centers <span class="op">=</span> [[<span class="dv">0</span>, <span class="dv">0</span>], [<span class="dv">5</span>, <span class="dv">5</span>]]</span>
<span id="cb131-3"><a href="#cb131-3"></a>X, labels_true <span class="op">=</span> make_blobs(</span>
<span id="cb131-4"><a href="#cb131-4"></a>    n_samples<span class="op">=</span><span class="dv">500</span>, centers<span class="op">=</span>centers, cluster_std<span class="op">=</span>[<span class="fl">2.0</span>,<span class="fl">0.5</span>], random_state<span class="op">=</span><span class="dv">0</span></span>
<span id="cb131-5"><a href="#cb131-5"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:4,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901567108,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="e5853bb6-dadb-40a7-cbf6-c3c4f9e5c660" data-execution_count="84">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1"></a>plt.scatter(X[:,<span class="dv">0</span>],X[:,<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="84">
<pre><code>&lt;matplotlib.collections.PathCollection at 0x7fdb95825810&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-79-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901567413,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="85">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1"></a>epss <span class="op">=</span> [<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>]</span>
<span id="cb134-2"><a href="#cb134-2"></a>min_sampless <span class="op">=</span> [<span class="dv">50</span>,<span class="dv">100</span>,<span class="dv">50</span>,<span class="dv">100</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:758,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901568805,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="14599b2b-cc27-419c-c33f-07298758a842" data-execution_count="86">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1"></a>f, axes <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">4</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">5</span>))</span>
<span id="cb135-2"><a href="#cb135-2"></a></span>
<span id="cb135-3"><a href="#cb135-3"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">4</span>):</span>
<span id="cb135-4"><a href="#cb135-4"></a>  db <span class="op">=</span> DBSCAN(eps<span class="op">=</span>epss[i], min_samples<span class="op">=</span>min_sampless[i]).fit(X)</span>
<span id="cb135-5"><a href="#cb135-5"></a>  y_pred <span class="op">=</span> db.labels_.astype(<span class="bu">int</span>)</span>
<span id="cb135-6"><a href="#cb135-6"></a>  colors <span class="op">=</span> np.array(</span>
<span id="cb135-7"><a href="#cb135-7"></a>            <span class="bu">list</span>(</span>
<span id="cb135-8"><a href="#cb135-8"></a>                islice(</span>
<span id="cb135-9"><a href="#cb135-9"></a>                    cycle(</span>
<span id="cb135-10"><a href="#cb135-10"></a>                        [</span>
<span id="cb135-11"><a href="#cb135-11"></a>                            <span class="st">"#377eb8"</span>,</span>
<span id="cb135-12"><a href="#cb135-12"></a>                            <span class="st">"#ff7f00"</span>,</span>
<span id="cb135-13"><a href="#cb135-13"></a>                            <span class="st">"#4daf4a"</span>,</span>
<span id="cb135-14"><a href="#cb135-14"></a>                            <span class="st">"#f781bf"</span>,</span>
<span id="cb135-15"><a href="#cb135-15"></a>                            <span class="st">"#a65628"</span>,</span>
<span id="cb135-16"><a href="#cb135-16"></a>                            <span class="st">"#984ea3"</span>,</span>
<span id="cb135-17"><a href="#cb135-17"></a>                            <span class="st">"#999999"</span>,</span>
<span id="cb135-18"><a href="#cb135-18"></a>                            <span class="st">"#e41a1c"</span>,</span>
<span id="cb135-19"><a href="#cb135-19"></a>                            <span class="st">"#dede00"</span>,</span>
<span id="cb135-20"><a href="#cb135-20"></a>                        ]</span>
<span id="cb135-21"><a href="#cb135-21"></a>                    ),</span>
<span id="cb135-22"><a href="#cb135-22"></a>                    <span class="bu">int</span>(<span class="bu">max</span>(y_pred) <span class="op">+</span> <span class="dv">1</span>),</span>
<span id="cb135-23"><a href="#cb135-23"></a>                )</span>
<span id="cb135-24"><a href="#cb135-24"></a>            )</span>
<span id="cb135-25"><a href="#cb135-25"></a>        )</span>
<span id="cb135-26"><a href="#cb135-26"></a>  <span class="co"># add black color for outliers (if any)</span></span>
<span id="cb135-27"><a href="#cb135-27"></a>  colors <span class="op">=</span> np.append(colors, [<span class="st">"#000000"</span>])</span>
<span id="cb135-28"><a href="#cb135-28"></a>  axes[i].scatter(X[:, <span class="dv">0</span>], X[:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span>colors[y_pred])</span>
<span id="cb135-29"><a href="#cb135-29"></a></span>
<span id="cb135-30"><a href="#cb135-30"></a>  axes[i].set_title(<span class="st">"eps=</span><span class="sc">%.1f</span><span class="st">, min_samples=</span><span class="sc">%d</span><span class="st">"</span> <span class="op">%</span> (epss[i],min_sampless[i]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-81-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>You can try to use HDBSCAN instead. For more information, please refer to <a href="https://hdbscan.readthedocs.io/en/latest/index.html">https://hdbscan.readthedocs.io/en/latest/index.html</a>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901569796,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="87">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1"></a>min_cluster_sizes <span class="op">=</span> [<span class="dv">5</span>,<span class="dv">10</span>,<span class="dv">30</span>,<span class="dv">50</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1036,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901571346,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="eb5f7df6-1167-4a02-b0ec-019b4c508a4c" data-execution_count="88">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb137-1"><a href="#cb137-1"></a>f, axes <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">4</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>, figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">5</span>))</span>
<span id="cb137-2"><a href="#cb137-2"></a></span>
<span id="cb137-3"><a href="#cb137-3"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">4</span>):</span>
<span id="cb137-4"><a href="#cb137-4"></a>  db <span class="op">=</span> hdbscan.HDBSCAN(min_cluster_size<span class="op">=</span>min_cluster_sizes[i]).fit(X)</span>
<span id="cb137-5"><a href="#cb137-5"></a>  y_pred <span class="op">=</span> db.labels_.astype(<span class="bu">int</span>)</span>
<span id="cb137-6"><a href="#cb137-6"></a>  colors <span class="op">=</span> np.array(</span>
<span id="cb137-7"><a href="#cb137-7"></a>            <span class="bu">list</span>(</span>
<span id="cb137-8"><a href="#cb137-8"></a>                islice(</span>
<span id="cb137-9"><a href="#cb137-9"></a>                    cycle(</span>
<span id="cb137-10"><a href="#cb137-10"></a>                        [</span>
<span id="cb137-11"><a href="#cb137-11"></a>                            <span class="st">"#377eb8"</span>,</span>
<span id="cb137-12"><a href="#cb137-12"></a>                            <span class="st">"#ff7f00"</span>,</span>
<span id="cb137-13"><a href="#cb137-13"></a>                            <span class="st">"#4daf4a"</span>,</span>
<span id="cb137-14"><a href="#cb137-14"></a>                            <span class="st">"#f781bf"</span>,</span>
<span id="cb137-15"><a href="#cb137-15"></a>                            <span class="st">"#a65628"</span>,</span>
<span id="cb137-16"><a href="#cb137-16"></a>                            <span class="st">"#984ea3"</span>,</span>
<span id="cb137-17"><a href="#cb137-17"></a>                            <span class="st">"#999999"</span>,</span>
<span id="cb137-18"><a href="#cb137-18"></a>                            <span class="st">"#e41a1c"</span>,</span>
<span id="cb137-19"><a href="#cb137-19"></a>                            <span class="st">"#dede00"</span>,</span>
<span id="cb137-20"><a href="#cb137-20"></a>                        ]</span>
<span id="cb137-21"><a href="#cb137-21"></a>                    ),</span>
<span id="cb137-22"><a href="#cb137-22"></a>                    <span class="bu">int</span>(<span class="bu">max</span>(y_pred) <span class="op">+</span> <span class="dv">1</span>),</span>
<span id="cb137-23"><a href="#cb137-23"></a>                )</span>
<span id="cb137-24"><a href="#cb137-24"></a>            )</span>
<span id="cb137-25"><a href="#cb137-25"></a>        )</span>
<span id="cb137-26"><a href="#cb137-26"></a>  <span class="co"># add black color for outliers (if any)</span></span>
<span id="cb137-27"><a href="#cb137-27"></a>  colors <span class="op">=</span> np.append(colors, [<span class="st">"#000000"</span>])</span>
<span id="cb137-28"><a href="#cb137-28"></a>  axes[i].scatter(X[:, <span class="dv">0</span>], X[:, <span class="dv">1</span>], s<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span>colors[y_pred])</span>
<span id="cb137-29"><a href="#cb137-29"></a>  axes[i].set_title(<span class="st">"min_cluster_size=</span><span class="sc">%d</span><span class="st">"</span> <span class="op">%</span> (min_cluster_sizes[i]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-83-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="nci60-data-example" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="nci60-data-example"><span class="header-section-number">10.3</span> NCI60 Data Example</h2>
<p>Unsupervised techniques are often used in the analysis of genomic data. In particular, PCA and hierarchical clustering are popular tools. We illustrate these techniques on the <code>NCI</code> cancer cell line microarray data, which consists of <span class="math inline">\(6{,}830\)</span> gene expression measurements on <span class="math inline">\(64\)</span> cancer cell lines.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1525,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901574932,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="ad9a2661-aa60-4ae4-a15c-a0811c4a0d63" data-execution_count="89">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1"></a>NCI60_labs <span class="op">=</span> pd.read_csv(<span class="st">"/content/drive/MyDrive/Lab/Data/NCI60_labs.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb138-2"><a href="#cb138-2"></a>NCI60_data <span class="op">=</span> pd.read_csv(<span class="st">"/content/drive/MyDrive/Lab/Data/NCI60_data.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb138-3"><a href="#cb138-3"></a></span>
<span id="cb138-4"><a href="#cb138-4"></a>NCI60_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="89">

  <div id="df-3f518798-30f1-4431-8656-34af5de6e54f">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
<th data-quarto-table-cell-role="th">4</th>
<th data-quarto-table-cell-role="th">5</th>
<th data-quarto-table-cell-role="th">6</th>
<th data-quarto-table-cell-role="th">7</th>
<th data-quarto-table-cell-role="th">8</th>
<th data-quarto-table-cell-role="th">9</th>
<th data-quarto-table-cell-role="th">10</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">6821</th>
<th data-quarto-table-cell-role="th">6822</th>
<th data-quarto-table-cell-role="th">6823</th>
<th data-quarto-table-cell-role="th">6824</th>
<th data-quarto-table-cell-role="th">6825</th>
<th data-quarto-table-cell-role="th">6826</th>
<th data-quarto-table-cell-role="th">6827</th>
<th data-quarto-table-cell-role="th">6828</th>
<th data-quarto-table-cell-role="th">6829</th>
<th data-quarto-table-cell-role="th">6830</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">V1</td>
<td>0.300000</td>
<td>1.180000</td>
<td>0.550000</td>
<td>1.140000</td>
<td>-0.265000</td>
<td>-7.000000e-02</td>
<td>0.350000</td>
<td>-0.315000</td>
<td>-0.450000</td>
<td>-0.654980</td>
<td>...</td>
<td>-0.990020</td>
<td>0.000000</td>
<td>0.030000</td>
<td>-0.175000</td>
<td>0.629981</td>
<td>-0.030000</td>
<td>0.000000</td>
<td>0.280000</td>
<td>-0.340000</td>
<td>-1.930000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">V2</td>
<td>0.679961</td>
<td>1.289961</td>
<td>0.169961</td>
<td>0.379961</td>
<td>0.464961</td>
<td>5.799610e-01</td>
<td>0.699961</td>
<td>0.724961</td>
<td>-0.040039</td>
<td>-0.285019</td>
<td>...</td>
<td>-0.270058</td>
<td>-0.300039</td>
<td>-0.250039</td>
<td>-0.535039</td>
<td>0.109941</td>
<td>-0.860039</td>
<td>-1.250049</td>
<td>-0.770039</td>
<td>-0.390039</td>
<td>-2.000039</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">V3</td>
<td>0.940000</td>
<td>-0.040000</td>
<td>-0.170000</td>
<td>-0.040000</td>
<td>-0.605000</td>
<td>0.000000e+00</td>
<td>0.090000</td>
<td>0.645000</td>
<td>0.430000</td>
<td>0.475019</td>
<td>...</td>
<td>0.319981</td>
<td>0.120000</td>
<td>-0.740000</td>
<td>-0.595000</td>
<td>-0.270020</td>
<td>-0.150000</td>
<td>0.000000</td>
<td>-0.120000</td>
<td>-0.410000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">V4</td>
<td>0.280000</td>
<td>-0.310000</td>
<td>0.680000</td>
<td>-0.810000</td>
<td>0.625000</td>
<td>-1.387779e-17</td>
<td>0.170000</td>
<td>0.245000</td>
<td>0.020000</td>
<td>0.095019</td>
<td>...</td>
<td>-1.240020</td>
<td>-0.110000</td>
<td>-0.160000</td>
<td>0.095000</td>
<td>-0.350019</td>
<td>-0.300000</td>
<td>-1.150010</td>
<td>1.090000</td>
<td>-0.260000</td>
<td>-1.100000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">V5</td>
<td>0.485000</td>
<td>-0.465000</td>
<td>0.395000</td>
<td>0.905000</td>
<td>0.200000</td>
<td>-5.000000e-03</td>
<td>0.085000</td>
<td>0.110000</td>
<td>0.235000</td>
<td>1.490019</td>
<td>...</td>
<td>0.554980</td>
<td>-0.775000</td>
<td>-0.515000</td>
<td>-0.320000</td>
<td>0.634980</td>
<td>0.605000</td>
<td>0.000000</td>
<td>0.745000</td>
<td>0.425000</td>
<td>0.145000</td>
</tr>
</tbody>
</table>


<p>5 rows  6830 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-3f518798-30f1-4431-8656-34af5de6e54f')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-3f518798-30f1-4431-8656-34af5de6e54f button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-3f518798-30f1-4431-8656-34af5de6e54f');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<p>Each cell line is labeled with a cancer type, given in <code>NCI60_labs</code>. We do not make use of the cancer types in performing PCA and clustering, as these are unsupervised techniques. But after performing PCA and clustering, we will check to see the extent to which these cancer types agree with the results of these unsupervised techniques.</p>
<p>The data has <span class="math inline">\(64\)</span> rows and <span class="math inline">\(6{,}830\)</span> columns.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901575503,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="39ddbe5b-ee01-4945-f0e1-bcc19015dd21" data-execution_count="90">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1"></a>NCI60_labs.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="90">
<pre><code>x          
NSCLC          9
RENAL          9
MELANOMA       8
BREAST         7
COLON          7
LEUKEMIA       6
OVARIAN        6
CNS            5
PROSTATE       2
K562A-repro    1
K562B-repro    1
MCF7A-repro    1
MCF7D-repro    1
UNKNOWN        1
dtype: int64</code></pre>
</div>
</div>
<section id="pca-on-the-nci60-data" class="level3" data-number="10.3.1">
<h3 data-number="10.3.1" class="anchored" data-anchor-id="pca-on-the-nci60-data"><span class="header-section-number">10.3.1</span> PCA on the NCI60 Data</h3>
<p>We first perform PCA on the data after scaling the variables (genes) to have standard deviation one, although one could reasonably argue that it is better not to scale the genes.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901576798,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="8d54c573-373f-4642-d240-b87c2c3e736f" data-execution_count="91">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1"></a>scaler <span class="op">=</span> StandardScaler()</span>
<span id="cb141-2"><a href="#cb141-2"></a>NCI60_scaled <span class="op">=</span> scaler.fit_transform(NCI60_data)</span>
<span id="cb141-3"><a href="#cb141-3"></a></span>
<span id="cb141-4"><a href="#cb141-4"></a>pca <span class="op">=</span> PCA()</span>
<span id="cb141-5"><a href="#cb141-5"></a>pca.fit(NCI60_scaled)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="91">
<pre><code>PCA()</code></pre>
</div>
</div>
<p>We now plot the first few principal component score vectors, in order to visualize the data. The observations (cell lines) corresponding to a given cancer type will be plotted in the same color, so that we can see to what extent the observations within a cancer type are similar to each other. We first create a simple function that assigns a distinct color to each element of a numeric vector. The function will be used to assign a color to each of the <span class="math inline">\(64\)</span> cell lines, based on the cancer type to which it corresponds.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901577680,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-execution_count="92">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb143-1"><a href="#cb143-1"></a>X_reduced <span class="op">=</span> pca.transform(NCI60_scaled)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:569,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901578548,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="f80ad78f-c2be-4212-bb62-0be053fe45d9" data-execution_count="93">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb144-1"><a href="#cb144-1"></a>fig, (ax1, ax2) <span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">1</span>, ncols<span class="op">=</span><span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">17</span>, <span class="dv">8</span>))</span>
<span id="cb144-2"><a href="#cb144-2"></a></span>
<span id="cb144-3"><a href="#cb144-3"></a>Z1 <span class="op">=</span> X_reduced[:, <span class="dv">0</span>]    </span>
<span id="cb144-4"><a href="#cb144-4"></a>Z2 <span class="op">=</span> X_reduced[:, <span class="dv">1</span>]   </span>
<span id="cb144-5"><a href="#cb144-5"></a>Z3 <span class="op">=</span> X_reduced[:, <span class="dv">2</span>]    </span>
<span id="cb144-6"><a href="#cb144-6"></a></span>
<span id="cb144-7"><a href="#cb144-7"></a><span class="co"># print(list(le.fit_transform(nci_labs.x)))</span></span>
<span id="cb144-8"><a href="#cb144-8"></a><span class="co"># list(enumerate(le.classes_))</span></span>
<span id="cb144-9"><a href="#cb144-9"></a>le <span class="op">=</span> LabelEncoder()</span>
<span id="cb144-10"><a href="#cb144-10"></a>ax1.scatter(Z1, Z2, s<span class="op">=</span><span class="dv">70</span>, c<span class="op">=</span>le.fit_transform(NCI60_labs), cmap<span class="op">=</span>plt.cm.gist_rainbow)     </span>
<span id="cb144-11"><a href="#cb144-11"></a>ax1.set_xlabel(<span class="st">'Z1'</span>, size<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb144-12"><a href="#cb144-12"></a>ax1.set_ylabel(<span class="st">'Z2'</span>, size<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb144-13"><a href="#cb144-13"></a></span>
<span id="cb144-14"><a href="#cb144-14"></a>ax2.scatter(Z1, Z3, s<span class="op">=</span><span class="dv">70</span>, c<span class="op">=</span>le.fit_transform(NCI60_labs), cmap<span class="op">=</span>plt.cm.gist_rainbow) </span>
<span id="cb144-15"><a href="#cb144-15"></a>ax2.set_xlabel(<span class="st">'Z1'</span>, size<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb144-16"><a href="#cb144-16"></a>ax2.set_ylabel(<span class="st">'Z3'</span>, size<span class="op">=</span><span class="dv">18</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:115: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="93">
<pre><code>Text(0, 0.5, 'Z3')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-88-output-3.png" class="img-fluid"></p>
</div>
</div>
<p>The resulting plots are shown in Figure 12.17. On the whole, cell lines corresponding to a single cancer type do tend to have similar values on the first few principal component score vectors. This indicates that cell lines from the same cancer type tend to have pretty similar gene expression levels.</p>
<p>We can obtain a summary of the proportion of variance explained (PVE) of the first few principal components:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2003,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901583757,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="9f6a9545-3cd0-40a3-d8c1-874927c7e24d" data-execution_count="94">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb147-1"><a href="#cb147-1"></a><span class="co"># Percentage of Variance Explained by each PC:</span></span>
<span id="cb147-2"><a href="#cb147-2"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">14</span>, <span class="dv">7</span>))</span>
<span id="cb147-3"><a href="#cb147-3"></a>plt.ylim(<span class="dv">0</span>, <span class="dv">100</span>)</span>
<span id="cb147-4"><a href="#cb147-4"></a>plt.xlim(<span class="op">-</span><span class="dv">1</span>, <span class="dv">65</span>)</span>
<span id="cb147-5"><a href="#cb147-5"></a>index <span class="op">=</span> np.arange(pca.components_.shape[<span class="dv">0</span>])</span>
<span id="cb147-6"><a href="#cb147-6"></a>plt.bar(index,  </span>
<span id="cb147-7"><a href="#cb147-7"></a>        height<span class="op">=</span>pca.explained_variance_ratio_ <span class="op">*</span> <span class="dv">100</span> , width<span class="op">=</span><span class="fl">0.3</span>, align<span class="op">=</span><span class="st">'center'</span>, </span>
<span id="cb147-8"><a href="#cb147-8"></a>        label<span class="op">=</span><span class="st">'percentage of variance explained'</span>)</span>
<span id="cb147-9"><a href="#cb147-9"></a>plt.plot(np.cumsum(pca.explained_variance_ratio_) <span class="op">*</span> <span class="dv">100</span>, <span class="st">'ro-'</span>, label<span class="op">=</span><span class="st">'cumulative PVE'</span>)</span>
<span id="cb147-10"><a href="#cb147-10"></a></span>
<span id="cb147-11"><a href="#cb147-11"></a>plt.ylabel(<span class="st">'percentage of variance explained'</span>, size<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb147-12"><a href="#cb147-12"></a>plt.xlabel(<span class="st">'Principal components'</span>, size<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb147-13"><a href="#cb147-13"></a>plt.legend(loc<span class="op">=</span><span class="st">'best'</span>, fontsize<span class="op">=</span><span class="st">'xx-large'</span>)</span>
<span id="cb147-14"><a href="#cb147-14"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-89-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We see that together, the first seven principal components explain around <span class="math inline">\(40\,\%\)</span> of the variance in the data. This is not a huge amount of the variance. However, looking at the scree plot, we see that while each of the first seven principal components explain a substantial amount of variance, there is a marked decrease in the variance explained by further principal components. That is, there is an <em>elbow</em> in the plot after approximately the seventh principal component. This suggests that there may be little benefit to examining more than seven or so principal components (though even examining seven principal components may be difficult).</p>
</section>
<section id="clustering-the-observations-of-the-nci60-data" class="level3" data-number="10.3.2">
<h3 data-number="10.3.2" class="anchored" data-anchor-id="clustering-the-observations-of-the-nci60-data"><span class="header-section-number">10.3.2</span> Clustering the Observations of the NCI60 Data</h3>
<p>We now proceed to hierarchically cluster the cell lines in the <code>NCI</code> data, with the goal of finding out whether or not the observations cluster into distinct types of cancer. To begin, we standardize the variables to have mean zero and standard deviation one. As mentioned earlier, this step is optional and should be performed only if we want each gene to be on the same <em>scale</em>.</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:3,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901584477,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="f54196c2-a725-464e-eaa0-4285972c841f" data-execution_count="95">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb148-1"><a href="#cb148-1"></a>NCI60_labs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="95">

  <div id="df-9ee3e3ed-a1d9-49b6-ad73-986fc40b1cf4">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">x</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>CNS</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>CNS</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>CNS</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>RENAL</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>BREAST</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">60</td>
<td>MELANOMA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">61</td>
<td>MELANOMA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">62</td>
<td>MELANOMA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">63</td>
<td>MELANOMA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">64</td>
<td>MELANOMA</td>
</tr>
</tbody>
</table>


<p>64 rows  1 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-9ee3e3ed-a1d9-49b6-ad73-986fc40b1cf4')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-9ee3e3ed-a1d9-49b6-ad73-986fc40b1cf4 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-9ee3e3ed-a1d9-49b6-ad73-986fc40b1cf4');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:8292,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901593228,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="89937d2f-86b6-4ac5-ef74-cfa024e57f98" data-execution_count="96">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb149-1"><a href="#cb149-1"></a>f, axes <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">1</span>, sharex<span class="op">=</span><span class="va">False</span>, sharey<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb149-2"><a href="#cb149-2"></a></span>
<span id="cb149-3"><a href="#cb149-3"></a>f.set_figheight(<span class="dv">24</span>)</span>
<span id="cb149-4"><a href="#cb149-4"></a>f.set_figwidth(<span class="dv">16</span>)</span>
<span id="cb149-5"><a href="#cb149-5"></a></span>
<span id="cb149-6"><a href="#cb149-6"></a>dendrogram(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'complete'</span>),</span>
<span id="cb149-7"><a href="#cb149-7"></a>           labels<span class="op">=</span>NCI60_labs.values,</span>
<span id="cb149-8"><a href="#cb149-8"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb149-9"><a href="#cb149-9"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb149-10"><a href="#cb149-10"></a>           ax<span class="op">=</span>axes[<span class="dv">0</span>])</span>
<span id="cb149-11"><a href="#cb149-11"></a></span>
<span id="cb149-12"><a href="#cb149-12"></a>dendrogram(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'average'</span>),</span>
<span id="cb149-13"><a href="#cb149-13"></a>           labels<span class="op">=</span>NCI60_labs.values,</span>
<span id="cb149-14"><a href="#cb149-14"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb149-15"><a href="#cb149-15"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb149-16"><a href="#cb149-16"></a>           ax<span class="op">=</span>axes[<span class="dv">1</span>])</span>
<span id="cb149-17"><a href="#cb149-17"></a></span>
<span id="cb149-18"><a href="#cb149-18"></a>dendrogram(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'single'</span>),</span>
<span id="cb149-19"><a href="#cb149-19"></a>           labels<span class="op">=</span>NCI60_labs.values,</span>
<span id="cb149-20"><a href="#cb149-20"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb149-21"><a href="#cb149-21"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>,</span>
<span id="cb149-22"><a href="#cb149-22"></a>           ax<span class="op">=</span>axes[<span class="dv">2</span>])</span>
<span id="cb149-23"><a href="#cb149-23"></a></span>
<span id="cb149-24"><a href="#cb149-24"></a>axes[<span class="dv">0</span>].set_title(<span class="st">'Complete Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb149-25"><a href="#cb149-25"></a>axes[<span class="dv">1</span>].set_title(<span class="st">'Average Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb149-26"><a href="#cb149-26"></a>axes[<span class="dv">2</span>].set_title(<span class="st">'Single Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="96">
<pre><code>Text(0.5, 1.0, 'Single Linkage')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-91-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>We see that the choice of linkage certainly does affect the results obtained. Typically, single linkage will tend to yield <em>trailing</em> clusters: very large clusters onto which individual observations attach one-by-one. On the other hand, complete and average linkage tend to yield more balanced, attractive clusters. For this reason, complete and average linkage are generally preferred to single linkage. Clearly cell lines within a single cancer type do tend to cluster together, although the clustering is not perfect. We will use complete linkage hierarchical clustering for the analysis that follows.</p>
<p>We can cut the dendrogram at the height that will yield a particular number of clusters, say four:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:728,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901593947,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="09f01483-99c5-496d-c5b2-b95152b82217" data-execution_count="97">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1"></a>cut_tree(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'complete'</span>), <span class="dv">4</span>).ravel()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="97">
<pre><code>array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 0, 3, 0,
       3, 3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])</code></pre>
</div>
</div>
<p>We can plot the cut on the dendrogram that produces these four clusters:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:1924,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901595870,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="cc6fd2c3-2195-455e-e796-1a2c5cd07db2" data-execution_count="98">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb153-1"><a href="#cb153-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">16</span>, <span class="dv">10</span>))</span>
<span id="cb153-2"><a href="#cb153-2"></a></span>
<span id="cb153-3"><a href="#cb153-3"></a>dendrogram(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'complete'</span>),</span>
<span id="cb153-4"><a href="#cb153-4"></a>           labels<span class="op">=</span>NCI60_labs.values,</span>
<span id="cb153-5"><a href="#cb153-5"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb153-6"><a href="#cb153-6"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb153-7"><a href="#cb153-7"></a></span>
<span id="cb153-8"><a href="#cb153-8"></a>plt.axhline(y<span class="op">=</span><span class="dv">139</span>, c<span class="op">=</span><span class="st">'r'</span>)</span>
<span id="cb153-9"><a href="#cb153-9"></a>plt.title(<span class="st">'Complete Linkage'</span>, size<span class="op">=</span><span class="dv">16</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-93-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We claimed earlier in Section 12.4.2 that <span class="math inline">\(K\)</span>-means clustering and hierarchical clustering with the dendrogram cut to obtain the same number of clusters can yield very different results. How do these <code>NCI</code> hierarchical clustering results compare to what we get if we perform <span class="math inline">\(K\)</span>-means clustering with <span class="math inline">\(K=4\)</span>?</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:830,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901596696,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="5da4c5a3-d8b3-40fe-e816-eda33175af9f" data-execution_count="99">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb154-1"><a href="#cb154-1"></a>kmeans <span class="op">=</span> KMeans(n_clusters<span class="op">=</span><span class="dv">4</span>, n_init<span class="op">=</span><span class="dv">20</span>, random_state<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb154-2"><a href="#cb154-2"></a>kmeans.fit(NCI60_scaled)</span>
<span id="cb154-3"><a href="#cb154-3"></a>pd.crosstab(kmeans.labels_, cut_tree(linkage(NCI60_scaled, method<span class="op">=</span><span class="st">'complete'</span>), <span class="dv">4</span>).ravel())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="99">

  <div id="df-0cc21e5c-d80b-423c-99e9-3be54f528c72">
    <div class="colab-df-container">
      <div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">col_0</th>
<th data-quarto-table-cell-role="th">0</th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">row_0</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>28</td>
<td>7</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>3</td>
<td>0</td>
<td>0</td>
<td>9</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>9</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0</td>
<td>0</td>
<td>8</td>
<td>0</td>
</tr>
</tbody>
</table>


</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-0cc21e5c-d80b-423c-99e9-3be54f528c72')" title="Convert this dataframe to an interactive table." style="display:none;">
        
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewbox="0 0 24 24" width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"></path><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"></path>
  </svg>
      </button>
      
  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-0cc21e5c-d80b-423c-99e9-3be54f528c72 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-0cc21e5c-d80b-423c-99e9-3be54f528c72');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
  
</div>
</div>
<p>We see that the four clusters obtained using hierarchical clustering and <span class="math inline">\(K\)</span>-means clustering are somewhat different. Cluster <span class="math inline">\(4\)</span> in <span class="math inline">\(K\)</span>-means clustering is identical to cluster <span class="math inline">\(3\)</span> in hierarchical clustering. However, the other clusters differ</p>
<p>Rather than performing hierarchical clustering on the entire data matrix, we can simply perform hierarchical clustering on the first few principal component score vectors, as follows:</p>
<div class="cell" data-executioninfo="{&quot;elapsed&quot;:2559,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1667901599252,&quot;user&quot;:{&quot;displayName&quot;:&quot;phonchi chung&quot;,&quot;userId&quot;:&quot;13517391734500420886&quot;},&quot;user_tz&quot;:-480}" data-outputid="f36937b7-f925-41dd-ce2c-54e1525508d1" data-execution_count="100">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb155-1"><a href="#cb155-1"></a>hc2 <span class="op">=</span> linkage(X_reduced[:, <span class="dv">0</span>:<span class="dv">5</span>], method<span class="op">=</span><span class="st">'complete'</span>)</span>
<span id="cb155-2"><a href="#cb155-2"></a></span>
<span id="cb155-3"><a href="#cb155-3"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">16</span>, <span class="dv">10</span>))</span>
<span id="cb155-4"><a href="#cb155-4"></a></span>
<span id="cb155-5"><a href="#cb155-5"></a>dendrogram(hc2,</span>
<span id="cb155-6"><a href="#cb155-6"></a>           labels<span class="op">=</span>NCI60_labs.values,</span>
<span id="cb155-7"><a href="#cb155-7"></a>           leaf_rotation<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb155-8"><a href="#cb155-8"></a>           leaf_font_size<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb155-9"><a href="#cb155-9"></a></span>
<span id="cb155-10"><a href="#cb155-10"></a>plt.title(<span class="st">'Hierarchical CLustering on First Five Score Vectors'</span>, size<span class="op">=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="100">
<pre><code>Text(0.5, 1.0, 'Hierarchical CLustering on First Five Score Vectors')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="Chapter_12_Lab_files/figure-html/cell-95-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>Not surprisingly, these results are different from the ones that we obtained when we performed hierarchical clustering on the full data set. Sometimes performing clustering on the first few principal component score vectors can give better results than performing clustering on the full data. In this situation, we might view the principal component step as one of denoising the data. We could also perform <span class="math inline">\(K\)</span>-means clustering on the first few principal component score vectors rather than the full data set.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./Chapter_9_Lab.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Support Vector Machines</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./NumPy_tutorial.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Numpy - multidimensional data arrays for python</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>